// mineral-ui/Form/FormFieldset.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = FormFieldset;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('../styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    FormFieldset_borderColor: baseTheme.borderColor,

    FormFieldsetLegend_color: baseTheme.h5_color,
    FormFieldsetLegend_fontSize: baseTheme.h5_fontSize,
    FormFieldsetLegend_fontWeight: baseTheme.h5_fontWeight

  }, baseTheme);
};

var Root = (0, _styles.createStyledComponent)('fieldset', function (_ref) {
  var disabled = _ref.disabled,
      baseTheme = _ref.theme;

  var theme = componentTheme(baseTheme);

  return {
    border: '1px solid transparent',
    borderTopColor: theme.FormFieldset_borderColor,
    padding: 0,

    '& > legend': {
      color: disabled ? theme.color_disabled : theme.FormFieldsetLegend_color,
      fontSize: theme.FormFieldsetLegend_fontSize,
      fontWeight: theme.FormFieldsetLegend_fontWeight,
      lineHeight: theme.size_medium,
      padding: 0,
      paddingLeft: theme.direction === 'rtl' ? theme.space_inline_sm : null,
      paddingRight: theme.direction === 'ltr' ? theme.space_inline_sm : null
    }
  };
}, {
  displayName: 'FormFieldset',
  includeStyleReset: true,
  rootEl: 'fieldset'
});

/**
 * FormFieldsets group related [FormFields](/components/form-field) and provide a legend.
 * Grouping FormFields provides contextual clues to users and enhances
 * accessibility.
 */
function FormFieldset(props) {
  var children = props.children,
      legend = props.legend,
      restProps = _objectWithoutProperties(props, ['children', 'legend']);

  return _react2.default.createElement(
    Root,
    restProps,
    legend && _jsx('legend', {}, void 0, legend),
    children
  );
}
FormFieldset.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Rendered content of the component, most likely [FormField](/components/form-field) */
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node) : _propTypes2.default.any).apply(this, arguments);
  },

  /** Disable (visually) the fieldset */
  disabled: _propTypes2.default.bool,

  /** Title of the fieldset */
  legend: _propTypes2.default.oneOfType([_propTypes2.default.string, function () {
    return (typeof React$Element === 'function' ? _propTypes2.default.instanceOf(React$Element) : _propTypes2.default.any).apply(this, arguments);
  }])
} : {};
