// @arranger/middleware/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AGGS_WRAPPER_NESTED = exports.AGGS_WRAPPER_FILTERED = exports.AGGS_WRAPPER_GLOBAL = exports.HISTOGRAM = exports.STATS = exports.BUCKETS = exports.ES_ARRANGER_SET_TYPE = exports.ES_ARRANGER_SET_INDEX = exports.ES_MAX_LONG = exports.ES_DATE_FORMAT = exports.ES_PHRASE_PREFIX = exports.ES_TYPE = exports.ES_FIELDS = exports.ES_MULTI_MATCH = exports.ES_PATH = exports.ES_QUERY = exports.ES_BOOL = exports.ES_PREFIX = exports.ES_NESTED = exports.ES_SHOULD = exports.ES_MUST_NOT = exports.ES_MUST = exports.ES_WILDCARD = exports.ARRAY_CONTENT = exports.DATE_FORMAT = exports.OP_ALIASES = exports.SET_ID = exports.MISSING = exports.REGEX = exports.BETWEEN_OP = exports.LTE_OP = exports.LT_OP = exports.GTE_OP = exports.GT_OP = exports.NOT_OP = exports.OR_OP = exports.AND_OP = exports.FILTER_OP = exports.SOME_NOT_IN_OP = exports.ALL_OP = exports.NOT_IN_OP = exports.IN_OP = void 0;
// sqon ops
var IN_OP = 'in';
exports.IN_OP = IN_OP;
var NOT_IN_OP = 'not-in';
exports.NOT_IN_OP = NOT_IN_OP;
var ALL_OP = 'all';
exports.ALL_OP = ALL_OP;
var SOME_NOT_IN_OP = 'some-not-in';
exports.SOME_NOT_IN_OP = SOME_NOT_IN_OP;
var FILTER_OP = 'filter';
exports.FILTER_OP = FILTER_OP;
var AND_OP = 'and';
exports.AND_OP = AND_OP;
var OR_OP = 'or';
exports.OR_OP = OR_OP;
var NOT_OP = 'not';
exports.NOT_OP = NOT_OP;
var GT_OP = 'gt';
exports.GT_OP = GT_OP;
var GTE_OP = 'gte';
exports.GTE_OP = GTE_OP;
var LT_OP = 'lt';
exports.LT_OP = LT_OP;
var LTE_OP = 'lte';
exports.LTE_OP = LTE_OP;
var BETWEEN_OP = 'between'; // special values

exports.BETWEEN_OP = BETWEEN_OP;
var REGEX = '*';
exports.REGEX = REGEX;
var MISSING = '__missing__';
exports.MISSING = MISSING;
var SET_ID = 'set_id:'; // sqon op aliases

exports.SET_ID = SET_ID;
var OP_ALIASES = {
  '>': GT_OP,
  '<': LT_OP,
  '>=': GTE_OP,
  '<=': LTE_OP,
  '=': IN_OP,
  '!=': NOT_IN_OP
};
exports.OP_ALIASES = OP_ALIASES;
var DATE_FORMAT = 'YYYY-MM-DD';
exports.DATE_FORMAT = DATE_FORMAT;
var ARRAY_CONTENT = [IN_OP, NOT_IN_OP, SOME_NOT_IN_OP, GT_OP, GTE_OP, LT_OP, LTE_OP]; // elasticsearch values

exports.ARRAY_CONTENT = ARRAY_CONTENT;
var ES_WILDCARD = 'wildcard';
exports.ES_WILDCARD = ES_WILDCARD;
var ES_MUST = 'must';
exports.ES_MUST = ES_MUST;
var ES_MUST_NOT = 'must_not';
exports.ES_MUST_NOT = ES_MUST_NOT;
var ES_SHOULD = 'should';
exports.ES_SHOULD = ES_SHOULD;
var ES_NESTED = 'nested';
exports.ES_NESTED = ES_NESTED;
var ES_PREFIX = 'prefix';
exports.ES_PREFIX = ES_PREFIX;
var ES_BOOL = 'bool';
exports.ES_BOOL = ES_BOOL;
var ES_QUERY = 'query';
exports.ES_QUERY = ES_QUERY;
var ES_PATH = 'path';
exports.ES_PATH = ES_PATH;
var ES_MULTI_MATCH = 'multi_match';
exports.ES_MULTI_MATCH = ES_MULTI_MATCH;
var ES_FIELDS = 'fields';
exports.ES_FIELDS = ES_FIELDS;
var ES_TYPE = 'type';
exports.ES_TYPE = ES_TYPE;
var ES_PHRASE_PREFIX = 'phrase_prefix';
exports.ES_PHRASE_PREFIX = ES_PHRASE_PREFIX;
var ES_DATE_FORMAT = 'YYYY-MM-DD HH:mm:ss.SSSSSS';
exports.ES_DATE_FORMAT = ES_DATE_FORMAT;
var ES_MAX_LONG = "-9223372036854775808";
exports.ES_MAX_LONG = ES_MAX_LONG;
var ES_ARRANGER_SET_INDEX = 'arranger-sets';
exports.ES_ARRANGER_SET_INDEX = ES_ARRANGER_SET_INDEX;
var ES_ARRANGER_SET_TYPE = 'arranger-sets';
exports.ES_ARRANGER_SET_TYPE = ES_ARRANGER_SET_TYPE;
var BUCKETS = 'buckets';
exports.BUCKETS = BUCKETS;
var STATS = 'stats';
exports.STATS = STATS;
var HISTOGRAM = 'histogram';
exports.HISTOGRAM = HISTOGRAM;
var AGGS_WRAPPER_GLOBAL = 'global';
exports.AGGS_WRAPPER_GLOBAL = AGGS_WRAPPER_GLOBAL;
var AGGS_WRAPPER_FILTERED = 'filtered';
exports.AGGS_WRAPPER_FILTERED = AGGS_WRAPPER_FILTERED;
var AGGS_WRAPPER_NESTED = 'nested';
exports.AGGS_WRAPPER_NESTED = AGGS_WRAPPER_NESTED;
