// @arranger/admin/dist/schemas/ExtendedMapping/schemaTypeDefs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const apollo_server_1 = require("apollo-server");
const convert = require("convert-units");
exports.default = () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    return apollo_server_1.gql `
  scalar JSON
  enum ExtendedFieldType {
    string
    object
    text
    boolean
    date
    keyword
    id
    long
    double
    integer
    float
    nested
  }
  # enum NumericTypeUnit {
  #   ${convert().measures()}
  # }

  type ExtendedFieldMapping {
    gqlId: ID!
    field: String!
    type: ExtendedFieldType!
    displayName: String!
    active: Boolean!
    isArray: Boolean!
    primaryKey: Boolean!
    quickSearchEnabled: Boolean!
    unit: String
    displayValues: JSON!
    rangeStep: Float
  }
  type Query {
    extendedFieldMappings(
      projectId: String!
      graphqlField: String!
      field: String
    ): [ExtendedFieldMapping]
  }

  input ExtendedFieldMappingInput {
    type: ExtendedFieldType
    displayName: String!
    active: Boolean!
    isArray: Boolean!
    primaryKey: Boolean!
    quickSearchEnabled: Boolean!
    unit: String
    displayValues: JSON
    rangeStep: Float
  }

  input ExtendedMappingSetFieldInput {
    field: String!
    type: ExtendedFieldType
    displayName: String!
    active: Boolean!
    isArray: Boolean!
    primaryKey: Boolean!
    quickSearchEnabled: Boolean!
    unit: String
    displayValues: JSON
    rangeStep: Float
  }

  type Mutation {
    updateExtendedMapping(
      projectId: String!
      graphqlField: String!
      field: String!
      extendedFieldMappingInput: ExtendedFieldMappingInput!
    ): ExtendedFieldMapping
    saveExtendedMapping(
      projectId: String!
      graphqlField: String!
      input: [ExtendedMappingSetFieldInput]!
    ): [ExtendedFieldMapping]
  }
`;
});
