// @arranger/admin/dist/schemas/ExtendedMapping/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var E_ExtendedFieldType;
(function (E_ExtendedFieldType) {
    E_ExtendedFieldType["string"] = "string";
    E_ExtendedFieldType["object"] = "object";
    E_ExtendedFieldType["text"] = "text";
    E_ExtendedFieldType["boolean"] = "boolean";
    E_ExtendedFieldType["date"] = "date";
    E_ExtendedFieldType["keyword"] = "keyword";
    E_ExtendedFieldType["id"] = "id";
    E_ExtendedFieldType["long"] = "long";
    E_ExtendedFieldType["double"] = "double";
    E_ExtendedFieldType["integer"] = "integer";
    E_ExtendedFieldType["float"] = "float";
    E_ExtendedFieldType["nested"] = "nested";
})(E_ExtendedFieldType || (E_ExtendedFieldType = {}));
var E_NumericTypeUnit;
(function (E_NumericTypeUnit) {
    E_NumericTypeUnit["length"] = "length";
    E_NumericTypeUnit["area"] = "area";
    E_NumericTypeUnit["mass"] = "mass";
    E_NumericTypeUnit["volume"] = "volume";
    E_NumericTypeUnit["each"] = "each";
    E_NumericTypeUnit["temperature"] = "temperature";
    E_NumericTypeUnit["time"] = "time";
    E_NumericTypeUnit["digital"] = "digital";
    E_NumericTypeUnit["partsPer"] = "partsPer";
    E_NumericTypeUnit["speed"] = "speed";
    E_NumericTypeUnit["pace"] = "pace";
    E_NumericTypeUnit["pressure"] = "pressure";
    E_NumericTypeUnit["current"] = "current";
    E_NumericTypeUnit["voltage"] = "voltage";
    E_NumericTypeUnit["power"] = "power";
    E_NumericTypeUnit["reactivePower"] = "reactivePower";
    E_NumericTypeUnit["apparentPower"] = "apparentPower";
    E_NumericTypeUnit["energy"] = "energy";
    E_NumericTypeUnit["reactiveEnergy"] = "reactiveEnergy";
    E_NumericTypeUnit["volumeFlowRate"] = "volumeFlowRate";
    E_NumericTypeUnit["illuminance"] = "illuminance";
    E_NumericTypeUnit["frequency"] = "frequency";
    E_NumericTypeUnit["angle"] = "angle";
})(E_NumericTypeUnit || (E_NumericTypeUnit = {}));
