// mineral-ui/es/Grid/GridItem.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import React from 'react';
import { createStyledComponent, getResponsiveStyles } from '../styles';
import { FlexItem } from '../Flex';


var getFlexGrow = function getFlexGrow(value) {
  return value ? 0 : 1;
};
var getWidth = function getWidth(value, columns, gutter) {
  return value ? 'calc(' + value / columns * 100 + '% - ' + gutter + ')' : 0;
};

/*
 * [1] IE11 doesn't use the correct box-sizing model with the flex-basis
 *     property. The workaround is to set flex-basis to 'auto' and use 'width'
 *     instead.
 */
var styles = {
  root: function root(_ref) {
    var breakpoints = _ref.breakpoints,
        columns = _ref.columns,
        gutterWidth = _ref.gutterWidth,
        span = _ref.span,
        theme = _ref.theme;

    var gutter = typeof gutterWidth === 'number' ? gutterWidth + 'px' : theme['space_inline_' + gutterWidth] || gutterWidth;

    var mapValueToProperty = function mapValueToProperty(property, value) {
      var map = {
        flexGrow: getFlexGrow,
        width: function width(value) {
          return getWidth(value, columns, gutter);
        }
      };

      return map[property](value);
    };

    return _extends({
      flexBasis: 'auto' }, getResponsiveStyles({
      breakpoints: breakpoints,
      mapValueToProperty: mapValueToProperty,
      styles: {
        flexGrow: span,
        width: span // [1]
      },
      theme: theme
    }));
  }
};

var Root = createStyledComponent(FlexItem, styles.root, {
  displayName: 'GridItem',
  filterProps: ['alignSelf', 'grow', 'inline', 'width'],
  withProps: { shrink: 0 }
});

/**
 * GridItem is used within [Grid](/components/grid) to lay out other components
 * in your app.
 */
export default function GridItem(props) {
  return React.createElement(Root, props);
}
GridItem.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Number of columns spanned [[Responsive-capable]](#responsive) */
  span: PropTypes.oneOfType([PropTypes.number, PropTypes.arrayOf(PropTypes.number)])
} : {};
import { bpfrpt_proptype_Values } from '../Flex/Flex';
import PropTypes from 'prop-types';
