// mineral-ui/utils/isValidProp.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _reactHtmlAttributes = require('react-html-attributes');

var _reactHtmlAttributes2 = _interopRequireDefault(_reactHtmlAttributes);

var _fastMemoize = require('fast-memoize');

var _fastMemoize2 = _interopRequireDefault(_fastMemoize);

var _reactProps = require('./reactProps');

var _reactProps2 = _interopRequireDefault(_reactProps);

var _collections = require('./collections');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var REGEX_DATA_OR_ARIA = /^(data|aria)-/;
var htmlAttributes = (0, _collections.settify)(_reactHtmlAttributes2.default);
var globalHtmlAttributes = htmlAttributes['*'] || new Set();
var svgTags = htmlAttributes.elements ? htmlAttributes.elements.svg : new Set();
var svgAttributes = htmlAttributes.svg || new Set();

var isElement = function isElement(tag) {
  return typeof tag === 'string';
};
var isComponent = function isComponent(tag) {
  return !isElement(tag);
};
var isReactProp = function isReactProp(prop) {
  return _reactProps2.default[prop];
};
var isCustomAttribute = function isCustomAttribute(prop) {
  return REGEX_DATA_OR_ARIA.test(prop);
};
var isSvgAttribute = function isSvgAttribute(tag, prop) {
  return svgTags.has(tag) && svgAttributes.has(prop);
};
var isHtmlAttribute = function isHtmlAttribute(tag, prop) {
  return globalHtmlAttributes.has(prop) || htmlAttributes[tag] && htmlAttributes[tag].has(prop);
};

var isValidProp = function isValidProp(tag, prop) {
  return isComponent(tag) || isReactProp(prop) || isHtmlAttribute(tag, prop) || isSvgAttribute(tag, prop) || isCustomAttribute(prop);
};

exports.default = (0, _fastMemoize2.default)(isValidProp);
