// @arranger/mapping-utils/dist/index.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "addMappingsToTypes", {
  enumerable: true,
  get: function get() {
    return _addMappingsToTypes["default"];
  }
});
Object.defineProperty(exports, "createConnectionResolvers", {
  enumerable: true,
  get: function get() {
    return _createConnectionResolvers["default"];
  }
});
Object.defineProperty(exports, "resolveSets", {
  enumerable: true,
  get: function get() {
    return _resolveSets["default"];
  }
});
Object.defineProperty(exports, "saveSet", {
  enumerable: true,
  get: function get() {
    return _resolveSets.saveSet;
  }
});
Object.defineProperty(exports, "mappingToFields", {
  enumerable: true,
  get: function get() {
    return _mappingToFields["default"];
  }
});
Object.defineProperty(exports, "mappingToAggsType", {
  enumerable: true,
  get: function get() {
    return _mappingToAggsType["default"];
  }
});
Object.defineProperty(exports, "mappingToAggsState", {
  enumerable: true,
  get: function get() {
    return _mappingToAggsState["default"];
  }
});
Object.defineProperty(exports, "esToAggTypeMap", {
  enumerable: true,
  get: function get() {
    return _esToAggTypeMap["default"];
  }
});
Object.defineProperty(exports, "mappingToColumnsState", {
  enumerable: true,
  get: function get() {
    return _mappingToColumnsState["default"];
  }
});
Object.defineProperty(exports, "mappingToMatchBoxState", {
  enumerable: true,
  get: function get() {
    return _mappingToMatchBoxState["default"];
  }
});
Object.defineProperty(exports, "mappingToNestedTypes", {
  enumerable: true,
  get: function get() {
    return _mappingToNestedTypes["default"];
  }
});
Object.defineProperty(exports, "mappingToScalarFields", {
  enumerable: true,
  get: function get() {
    return _mappingToScalarFields["default"];
  }
});
Object.defineProperty(exports, "esToGraphqlTypeMap", {
  enumerable: true,
  get: function get() {
    return _mappingToScalarFields.esToGraphqlTypeMap;
  }
});
Object.defineProperty(exports, "fetchMapping", {
  enumerable: true,
  get: function get() {
    return _fetchMapping.fetchMapping;
  }
});
Object.defineProperty(exports, "loadExtendedFields", {
  enumerable: true,
  get: function get() {
    return _loadExtendedFields["default"];
  }
});
Object.defineProperty(exports, "mapHits", {
  enumerable: true,
  get: function get() {
    return _mapHits["default"];
  }
});
Object.defineProperty(exports, "getNestedFields", {
  enumerable: true,
  get: function get() {
    return _getNestedFields["default"];
  }
});
Object.defineProperty(exports, "flattenMapping", {
  enumerable: true,
  get: function get() {
    return _flattenMapping["default"];
  }
});
Object.defineProperty(exports, "extendMapping", {
  enumerable: true,
  get: function get() {
    return _extendMapping["default"];
  }
});
Object.defineProperty(exports, "extendFields", {
  enumerable: true,
  get: function get() {
    return _extendMapping.extendFields;
  }
});
Object.defineProperty(exports, "mappingToDisplayTreeData", {
  enumerable: true,
  get: function get() {
    return _mappingToDisplayTreeData["default"];
  }
});

var _addMappingsToTypes = _interopRequireDefault(require("./addMappingsToTypes"));

var _createConnectionResolvers = _interopRequireDefault(require("./createConnectionResolvers"));

var _resolveSets = _interopRequireWildcard(require("./resolveSets"));

var _mappingToFields = _interopRequireDefault(require("./mappingToFields"));

var _mappingToAggsType = _interopRequireDefault(require("./mappingToAggsType"));

var _mappingToAggsState = _interopRequireDefault(require("./mappingToAggsState"));

var _esToAggTypeMap = _interopRequireDefault(require("./esToAggTypeMap"));

var _mappingToColumnsState = _interopRequireDefault(require("./mappingToColumnsState"));

var _mappingToMatchBoxState = _interopRequireDefault(require("./mappingToMatchBoxState"));

var _mappingToNestedTypes = _interopRequireDefault(require("./mappingToNestedTypes"));

var _mappingToScalarFields = _interopRequireWildcard(require("./mappingToScalarFields"));

var _fetchMapping = require("./utils/fetchMapping");

var _loadExtendedFields = _interopRequireDefault(require("./utils/loadExtendedFields"));

var _mapHits = _interopRequireDefault(require("./utils/mapHits"));

var _getNestedFields = _interopRequireDefault(require("./getNestedFields"));

var _flattenMapping = _interopRequireDefault(require("./flattenMapping"));

var _extendMapping = _interopRequireWildcard(require("./extendMapping"));

var _mappingToDisplayTreeData = _interopRequireDefault(require("./mappingToDisplayTreeData"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
