// mineral-ui-tokens/es/all.js
export var backgroundColor = '#ffffff';
export var backgroundColor_active = '#ebeff5';
export var backgroundColor_disabled = '#ebeff5';
export var backgroundColor_focus = '#ffffff';
export var backgroundColor_hover = '#f5f7fa';
export var backgroundColor_brand_selected = '#f0f5fc';
export var backgroundColor_brand_selectedActive = '#accbfc';
export var backgroundColor_brand_selectedHover = '#cfe0fc';
export var backgroundColor_brandPrimary = '#3272d9';
export var backgroundColor_brandPrimary_active = '#1d5bbf';
export var backgroundColor_brandPrimary_focus = '#3272d9';
export var backgroundColor_brandPrimary_hover = '#5691f0';
export var backgroundColor_danger_active = '#fad4d4';
export var backgroundColor_danger_focus = '#faf0f0';
export var backgroundColor_danger_hover = '#faf0f0';
export var backgroundColor_dangerPrimary = '#de1b1b';
export var backgroundColor_dangerPrimary_active = '#b80d0d';
export var backgroundColor_dangerPrimary_focus = '#de1b1b';
export var backgroundColor_dangerPrimary_hover = '#f55353';
export var backgroundColor_success_active = '#abedc5';
export var backgroundColor_success_focus = '#e1faeb';
export var backgroundColor_success_hover = '#e1faeb';
export var backgroundColor_successPrimary = '#2a854e';
export var backgroundColor_successPrimary_active = '#20693d';
export var backgroundColor_successPrimary_focus = '#2a854e';
export var backgroundColor_successPrimary_hover = '#3ba164';
export var backgroundColor_warning_active = '#fad8af';
export var backgroundColor_warning_focus = '#fcf2e6';
export var backgroundColor_warning_hover = '#fcf2e6';
export var backgroundColor_warningPrimary = '#ad5f00';
export var backgroundColor_warningPrimary_active = '#8a4d03';
export var backgroundColor_warningPrimary_focus = '#ad5f00';
export var backgroundColor_warningPrimary_hover = '#cf7911';
export var borderColor = '#c8d1e0';
export var borderColor_brand = '#3272d9';
export var borderColor_brand_active = '#1d5bbf';
export var borderColor_brand_focus = '#1d5bbf';
export var borderColor_brand_hover = '#5691f0';
export var borderColor_danger = '#de1b1b';
export var borderColor_danger_active = '#b80d0d';
export var borderColor_danger_focus = '#b80d0d';
export var borderColor_danger_hover = '#f55353';
export var borderColor_success = '#2a854e';
export var borderColor_success_active = '#20693d';
export var borderColor_success_focus = '#20693d';
export var borderColor_success_hover = '#3ba164';
export var borderColor_warning = '#ad5f00';
export var borderColor_warning_active = '#8a4d03';
export var borderColor_warning_focus = '#8a4d03';
export var borderColor_warning_hover = '#cf7911';
export var borderRadius_1 = '3px';
export var boxShadow_1 = '0 1px 2px 0 rgba(0,0,0,0.2), 0 2px 4px 0 rgba(0,0,0,0.2)';
export var boxShadow_2 = '0 2px 4px 0 rgba(0,0,0,0.2), 0 4px 8px 0 rgba(0,0,0,0.2)';
export var boxShadow_3 = '0 4px 8px 0 rgba(0,0,0,0.2), 0 8px 16px 0 rgba(0,0,0,0.2)';
export var boxShadow_4 = '0 8px 16px 0 rgba(0,0,0,0.2), 0 20px 16px -8px rgba(0,0,0,0.2)';
export var boxShadow_5 = '0 16px 24px 0 rgba(0,0,0,0.2), 0 32px 24px -16px rgba(0,0,0,0.2)';
export var breakpoint_narrow = '512px';
export var breakpoint_medium = '768px';
export var breakpoint_wide = '1024px';
export var color = '#333840';
export var color_inverted = '#ffffff';
export var color_disabled = '#afbacc';
export var color_mouse = '#58606e';
export var color_readOnly = '#58606e';
export var color_required = '#de1b1b';
export var color_brand = '#3272d9';
export var color_brandPrimary = '#ffffff';
export var color_brand_active = '#1d5bbf';
export var color_brand_focus = '#3272d9';
export var color_brand_hover = '#5691f0';
export var color_danger = '#de1b1b';
export var color_dangerPrimary = '#ffffff';
export var color_danger_active = '#b80d0d';
export var color_danger_focus = '#de1b1b';
export var color_danger_hover = '#f55353';
export var color_success = '#2a854e';
export var color_successPrimary = '#ffffff';
export var color_success_active = '#20693d';
export var color_success_focus = '#2a854e';
export var color_success_hover = '#3ba164';
export var color_warning = '#ad5f00';
export var color_warningPrimary = '#ffffff';
export var color_warning_active = '#8a4d03';
export var color_warning_focus = '#ad5f00';
export var color_warning_hover = '#cf7911';
export var fontFamily = '"Open Sans"';
export var fontFamily_system = '-apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"';
export var fontFamily_monospace = '"SF Mono", "Droid Sans Mono", "Source Code Pro", Monaco, Consolas, "Courier New", Courier, monospace';
export var fontSize_base = '1rem';
export var fontSize_mouse = '0.6875rem';
export var fontSize_prose = '1rem';
export var fontSize_ui = '0.875rem';
export var fontWeight_regular = 400;
export var fontWeight_semiBold = 600;
export var fontWeight_bold = 700;
export var fontWeight_extraBold = 800;
export var lineHeight = 1.25;
export var lineHeight_heading = 1.25;
export var lineHeight_heading_small = 1.5;
export var lineHeight_prose = 1.5;
export var size_small = '24px';
export var size_medium = '32px';
export var size_large = '40px';
export var size_jumbo = '52px';
export var space_inline_xxs = '2px';
export var space_inline_xs = '4px';
export var space_inline_sm = '8px';
export var space_inline_md = '16px';
export var space_inline_lg = '24px';
export var space_inline_xl = '32px';
export var space_inline_xxl = '64px';
export var space_inset_sm = '8px';
export var space_inset_md = '16px';
export var space_inset_lg = '24px';
export var space_stack_xxs = '2px';
export var space_stack_xs = '4px';
export var space_stack_sm = '8px';
export var space_stack_md = '16px';
export var space_stack_lg = '24px';
export var space_stack_xl = '32px';
export var space_stack_xxl = '64px';
export var zIndex_100 = 100;
export var zIndex_200 = 200;
export var zIndex_400 = 400;
export var zIndex_800 = 800;
export var zIndex_1600 = 1600;
export var h1_color = '#333840';
export var h1_fontSize = '2.125rem';
export var h1_fontWeight = 800;
export var h2_color = '#58606e';
export var h2_fontSize = '1.75rem';
export var h2_fontWeight = 700;
export var h3_color = '#58606e';
export var h3_fontSize = '1.375rem';
export var h3_fontWeight = 700;
export var h4_color = '#58606e';
export var h4_fontSize = '1.125rem';
export var h4_fontWeight = 700;
export var h5_color = '#333840';
export var h5_fontSize = '0.875rem';
export var h5_fontWeight = 700;
export var h6_color = '#58606e';
export var h6_fontSize = '0.875rem';
export var h6_fontWeight = 400;
export var icon_color = '#58606e';
export var icon_color_brand = '#3272d9';
export var icon_color_danger = '#de1b1b';
export var icon_color_success = '#2a854e';
export var icon_color_warning = '#ad5f00';
export var input_backgroundColor = '#ffffff';
export var input_backgroundColor_disabled = '#ebeff5';
export var input_color_placeholder = '#8e99ab';
export var panel_backgroundColor = '#ffffff';
export var panel_backgroundColor_inverted = '#434a54';
export var panel_borderColor = '#ebeff5';
export var panel_borderColor_inverted = '#434a54';
export var well_backgroundColor = '#ebeff5';
export var well_backgroundColor_danger = '#fad4d4';
export var well_backgroundColor_success = '#abedc5';
export var well_backgroundColor_warning = '#fad8af';
export var well_borderColor_danger = '#fa8e8e';
export var well_borderColor_success = '#57c282';
export var well_borderColor_warning = '#e89c3f';
export var black = '#1d1f24';
export var white = '#ffffff';
export var brand_10 = '#f0f5fc';
export var brand_20 = '#cfe0fc';
export var brand_30 = '#accbfc';
export var brand_40 = '#84b1fa';
export var brand_50 = '#5691f0';
export var brand_60 = '#3272d9';
export var brand_70 = '#1d5bbf';
export var brand_80 = '#114599';
export var brand_90 = '#103570';
export var brand_100 = '#15233b';
export var blue_10 = '#f0f5fc';
export var blue_20 = '#cfe0fc';
export var blue_30 = '#accbfc';
export var blue_40 = '#84b1fa';
export var blue_50 = '#5691f0';
export var blue_60 = '#3272d9';
export var blue_70 = '#1d5bbf';
export var blue_80 = '#114599';
export var blue_90 = '#103570';
export var blue_100 = '#15233b';
export var bronze_10 = '#fcf2e6';
export var bronze_20 = '#fad8af';
export var bronze_30 = '#f5bc76';
export var bronze_40 = '#e89c3f';
export var bronze_50 = '#cf7911';
export var bronze_60 = '#ad5f00';
export var bronze_70 = '#8a4d03';
export var bronze_80 = '#693d07';
export var bronze_90 = '#4d2f0b';
export var bronze_100 = '#33210c';
export var dusk_10 = '#f4f2f7';
export var dusk_20 = '#e3dcf7';
export var dusk_30 = '#cec2f0';
export var dusk_40 = '#b5a6e3';
export var dusk_50 = '#9886cf';
export var dusk_60 = '#7a68b3';
export var dusk_70 = '#645396';
export var dusk_80 = '#4f4178';
export var dusk_90 = '#3a3154';
export var dusk_100 = '#282436';
export var gray_10 = '#f5f7fa';
export var gray_20 = '#ebeff5';
export var gray_30 = '#dde3ed';
export var gray_40 = '#c8d1e0';
export var gray_50 = '#afbacc';
export var gray_60 = '#8e99ab';
export var gray_70 = '#707a8a';
export var gray_80 = '#58606e';
export var gray_90 = '#434a54';
export var gray_100 = '#333840';
export var green_10 = '#e1faeb';
export var green_20 = '#abedc5';
export var green_30 = '#7ddba3';
export var green_40 = '#57c282';
export var green_50 = '#3ba164';
export var green_60 = '#2a854e';
export var green_70 = '#20693d';
export var green_80 = '#1a5230';
export var green_90 = '#153d25';
export var green_100 = '#112b1b';
export var indigo_10 = '#f2f2fc';
export var indigo_20 = '#dcdcfc';
export var indigo_30 = '#c2c2fc';
export var indigo_40 = '#a7a7fa';
export var indigo_50 = '#8585f2';
export var indigo_60 = '#6767e6';
export var indigo_70 = '#4d4dd1';
export var indigo_80 = '#3737b3';
export var indigo_90 = '#28288a';
export var indigo_100 = '#202057';
export var magenta_10 = '#faf0f4';
export var magenta_20 = '#fad4e4';
export var magenta_30 = '#fab4d1';
export var magenta_40 = '#f78bb8';
export var magenta_50 = '#ed5393';
export var magenta_60 = '#d6246e';
export var magenta_70 = '#b01355';
export var magenta_80 = '#8a1244';
export var magenta_90 = '#611535';
export var magenta_100 = '#421527';
export var purple_10 = '#f5f0fa';
export var purple_20 = '#ead9fa';
export var purple_30 = '#dabcf7';
export var purple_40 = '#c79bf2';
export var purple_50 = '#ae74e8';
export var purple_60 = '#9656d6';
export var purple_70 = '#7d3cbd';
export var purple_80 = '#642b9e';
export var purple_90 = '#4b2175';
export var purple_100 = '#371c52';
export var red_10 = '#faf0f0';
export var red_20 = '#fad4d4';
export var red_30 = '#fab6b6';
export var red_40 = '#fa8e8e';
export var red_50 = '#f55353';
export var red_60 = '#de1b1b';
export var red_70 = '#b80d0d';
export var red_80 = '#8f0e0e';
export var red_90 = '#661414';
export var red_100 = '#451717';
export var sky_10 = '#e8f4fa';
export var sky_20 = '#bbe5fa';
export var sky_30 = '#8dd4f7';
export var sky_40 = '#53baed';
export var sky_50 = '#229ad6';
export var sky_60 = '#0c7bb3';
export var sky_70 = '#066391';
export var sky_80 = '#064d70';
export var sky_90 = '#093952';
export var sky_100 = '#0c2938';
export var slate_10 = '#edf4f7';
export var slate_20 = '#cbe3f5';
export var slate_30 = '#a7ceeb';
export var slate_40 = '#84b7db';
export var slate_50 = '#5e95bd';
export var slate_60 = '#48799c';
export var slate_70 = '#376180';
export var slate_80 = '#2d4d63';
export var slate_90 = '#243947';
export var slate_100 = '#1d2830';
export var teal_10 = '#e4f7f6';
export var teal_20 = '#a8ede9';
export var teal_30 = '#6cd9d2';
export var teal_40 = '#45bfb7';
export var teal_50 = '#28a199';
export var teal_60 = '#17827b';
export var teal_70 = '#116963';
export var teal_80 = '#0d524d';
export var teal_90 = '#0b3d3a';
export var teal_100 = '#092b29';
