// mineral-ui/Popover/PopoverArrow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = PopoverArrow;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactPopper = require('react-popper');

var _styles = require('../styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    PopoverArrow_backgroundColor: baseTheme.panel_backgroundColor,
    PopoverArrow_borderColor: baseTheme.panel_borderColor
  }, baseTheme);
};

var Root = (0, _styles.createStyledComponent)(_reactPopper.Arrow, function (_ref) {
  var placement = _ref.placement,
      size = _ref.size,
      baseTheme = _ref.theme;

  var theme = componentTheme(baseTheme);
  var arrowShadow = ', 0 3px 1px rgba(0, 0, 0, 0.3)';
  var horizontalOffset = '-' + (parseFloat(size) - 4) + 'px';
  var directionalStyles = void 0;
  var rotation = 0;

  switch (true) {
    case placement && placement.startsWith('top'):
      // Magic numbers to optically match theme.boxShadow_2
      arrowShadow = ', 0 4px 2px rgba(0, 0, 0, 0.3)';
      directionalStyles = {
        bottom: '-' + (parseFloat(size) - 2) + 'px',
        left: 'calc(50% - ' + size + ')',
        marginBottom: 0,
        marginTop: 0
      };
      break;
    case placement && placement.startsWith('bottom'):
      arrowShadow = '';
      directionalStyles = {
        top: '-' + (parseFloat(size) - 3) + 'px',
        left: 'calc(50% - ' + size + ')',
        marginBottom: 0,
        marginTop: 0
      };
      rotation = 180;
      break;
    case placement && placement.startsWith('left'):
      directionalStyles = {
        right: horizontalOffset,
        top: 'calc(50% - ' + size + ')',
        marginLeft: 0,
        marginRight: 0
      };
      rotation = -90;
      break;
    case placement && placement.startsWith('right'):
      directionalStyles = {
        left: horizontalOffset,
        top: 'calc(50% - ' + size + ')',
        marginLeft: 0,
        marginRight: 0
      };
      rotation = 90;
      break;
    default:
      directionalStyles = {
        display: 'none'
      };
  }

  return _extends({
    color: theme.PopoverArrow_backgroundColor,
    display: 'inline-block',
    fontSize: size,
    margin: size,
    position: 'absolute',
    textShadow: '0 2px 0 ' + theme.PopoverArrow_borderColor + arrowShadow,
    transform: 'rotate(' + rotation + 'deg) scaleX(2)'
  }, directionalStyles);
}, {
  displayName: 'PopoverArrow',
  rootEl: 'span'
});

/**
 * PopoverArrow component
 */
function PopoverArrow(props) {
  return _react2.default.createElement(
    Root,
    _extends({}, props, { 'aria-hidden': true }),
    '\u25BC'
  );
}
PopoverArrow.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Size of arrow */
  size: _propTypes2.default.string.isRequired,

  /** Placement of the popper */
  placement: _propTypes2.default.oneOf(['auto', 'auto-end', 'auto-start', 'bottom', 'bottom-end', 'bottom-start', 'left', 'left-end', 'left-start', 'right', 'right-end', 'right-start', 'top', 'top-end', 'top-start'])
} : {};
