// mineral-ui/Dropdown/DropdownContent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('../styles');

var _RtlPopper = require('../Popover/RtlPopper');

var _RtlPopper2 = _interopRequireDefault(_RtlPopper);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    DropdownContent_backgroundColor: baseTheme.panel_backgroundColor,
    DropdownContent_borderColor: baseTheme.panel_borderColor,
    DropdownContent_borderRadius: baseTheme.borderRadius_1,
    DropdownContent_boxShadow: baseTheme.boxShadow_2,
    DropdownContent_margin: '5px',
    DropdownContent_zIndex: baseTheme.zIndex_100
  }, baseTheme);
};

var Root = (0, _styles.createStyledComponent)(_RtlPopper2.default, function (_ref) {
  var baseTheme = _ref.theme,
      wide = _ref.wide;

  var theme = componentTheme(baseTheme);

  return {
    backgroundColor: theme.DropdownContent_backgroundColor,
    border: '1px solid ' + theme.DropdownContent_borderColor,
    borderRadius: theme.DropdownContent_borderRadius,
    boxShadow: theme.DropdownContent_boxShadow,
    maxHeight: '30vh',
    overflowY: 'auto',
    userSelect: 'none',
    width: wide ? (0, _styles.pxToEm)(344) : (0, _styles.pxToEm)(224),
    zIndex: theme.DropdownContent_zIndex,

    '&[data-placement^="top"]': {
      marginBottom: theme.DropdownContent_margin
    },
    '&[data-placement^="bottom"]': {
      marginTop: theme.DropdownContent_margin
    },
    '&[data-placement^="left"]': {
      marginRight: theme.DropdownContent_margin
    },
    '&[data-placement^="right"]': {
      marginLeft: theme.DropdownContent_margin
    },
    '&[data-x-out-of-boundaries]': {
      visibility: 'hidden'
    }
  };
}, {
  displayName: 'DropdownContent',
  includeStyleReset: true,
  filterProps: ['hasArrow', 'wide']
});

/**
 * DropdownContent component
 */

var DropdownContent = function (_Component) {
  _inherits(DropdownContent, _Component);

  function DropdownContent() {
    _classCallCheck(this, DropdownContent);

    return _possibleConstructorReturn(this, (DropdownContent.__proto__ || Object.getPrototypeOf(DropdownContent)).apply(this, arguments));
  }

  _createClass(DropdownContent, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          rootProps = _objectWithoutProperties(_props, ['children']);

      return _react2.default.createElement(
        Root,
        rootProps,
        function (_ref2) {
          var popperProps = _ref2.popperProps,
              restProps = _ref2.restProps;

          var wrapperProps = _extends({}, popperProps, restProps);

          return _react2.default.createElement(
            'div',
            wrapperProps,
            children
          );
        }
      );
    }
  }]);

  return DropdownContent;
}(_react.Component);

exports.default = DropdownContent;
DropdownContent.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Content of the Dropdown */
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node).isRequired : _propTypes2.default.any.isRequired).apply(this, arguments);
  },

  /** Id of the Dropdown content */
  id: _propTypes2.default.string.isRequired,

  /** Plugins that are used to alter behavior. See https://popper.js.org/popper-documentation.html#modifiers */
  modifiers: _propTypes2.default.object,

  /** Placement of the dropdown */
  placement: _propTypes2.default.oneOf(['bottom-end', 'bottom-start', 'left-end', 'left-start', 'right-end', 'right-start', 'top-end', 'top-start']),

  /** Display a wider dropdown menu */
  wide: _propTypes2.default.bool
} : {};
