// mineral-ui/Dialog/Dialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _exenv = require('exenv');

var _focusTrapReact = require('focus-trap-react');

var _focusTrapReact2 = _interopRequireDefault(_focusTrapReact);

var _noScroll = require('no-scroll');

var _noScroll2 = _interopRequireDefault(_noScroll);

var _Transition = require('react-transition-group/Transition');

var _Transition2 = _interopRequireDefault(_Transition);

var _styles = require('../styles');

var _themes = require('../themes');

var _utils = require('../utils');

var _children = require('../utils/children');

var _Button = require('../Button');

var _Button2 = _interopRequireDefault(_Button);

var _IconClose = require('../Icon/IconClose');

var _IconClose2 = _interopRequireDefault(_IconClose);

var _Portal = require('../Portal');

var _Portal2 = _interopRequireDefault(_Portal);

var _EventListener = require('../EventListener');

var _EventListener2 = _interopRequireDefault(_EventListener);

var _DialogActions = require('./DialogActions');

var _DialogActions2 = _interopRequireDefault(_DialogActions);

var _DialogBody = require('./DialogBody');

var _DialogBody2 = _interopRequireDefault(_DialogBody);

var _DialogFooter = require('./DialogFooter');

var _DialogFooter2 = _interopRequireDefault(_DialogFooter);

var _DialogHeader = require('./DialogHeader');

var _DialogHeader2 = _interopRequireDefault(_DialogHeader);

var _DialogTitle = require('./DialogTitle');

var _DialogTitle2 = _interopRequireDefault(_DialogTitle);

var _DialogRow = require('./DialogRow');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    Dialog_transitionDuration: '250ms',
    Dialog_zIndex: baseTheme.zIndex_1600,

    DialogCloseButton_margin: baseTheme.space_inline_sm,

    DialogContent_backgroundColor: baseTheme.panel_backgroundColor,
    DialogContent_borderColor: baseTheme.panel_borderColor,
    DialogContent_borderRadius: baseTheme.borderRadius_1,
    DialogContent_boxShadow: baseTheme.boxShadow_5,
    DialogContent_maxHeight: '80vh',
    DialogContent_maxHeight_small: (0, _styles.pxToEm)(560),
    DialogContent_maxHeight_medium: (0, _styles.pxToEm)(560),
    DialogContent_maxHeight_large: (0, _styles.pxToEm)(720),
    DialogContent_maxWidth_small: (0, _styles.pxToEm)(400),
    DialogContent_maxWidth_medium: (0, _styles.pxToEm)(640),
    DialogContent_maxWidth_large: (0, _styles.pxToEm)(1200),
    DialogContent_minWidth: (0, _styles.pxToEm)(360),
    DialogContent_offsetVertical: baseTheme.space_stack_xxl,
    DialogContent_width_small: '35vw',
    DialogContent_width_medium: '50vw',
    DialogContent_width_large: '80vw',

    DialogOverlay_backgroundColor: 'rgba(0, 0, 0, 0.6)'

  }, (0, _DialogRow.componentTheme)(baseTheme), baseTheme);
};

var styles = {
  animate: function animate(_ref) {
    var state = _ref.state,
        baseTheme = _ref.theme;

    var theme = componentTheme(baseTheme);
    return {
      opacity: state === 'entered' ? 1 : 0,
      position: 'relative',
      transition: 'opacity ' + theme.Dialog_transitionDuration + ' ease',
      willChange: 'opacity',
      zIndex: theme.Dialog_zIndex
    };
  },
  content: function content(_ref2) {
    var size = _ref2.size,
        baseTheme = _ref2.theme;

    var theme = componentTheme(baseTheme);

    var getSizeStyles = function getSizeStyles(size) {
      var maxWidth = theme['DialogContent_maxWidth_' + size];
      var maxHeight = theme['DialogContent_maxHeight_' + size];
      var width = theme['DialogContent_width_' + size];
      var offsetVertical = theme.DialogContent_offsetVertical;

      var maxHeightNumber = parseFloat(maxHeight);
      var offsetVerticalNumber = parseFloat(offsetVertical);
      var minHeight = maxHeightNumber + 2 * offsetVerticalNumber + 'em';

      return _defineProperty({
        maxWidth: maxWidth,
        width: width

      }, '@media(min-height: ' + minHeight + ')', {
        maxHeight: maxHeight
      });
    };

    return _extends({
      backgroundColor: theme.DialogContent_backgroundColor,
      border: '1px solid ' + theme.DialogContent_borderColor,
      borderRadius: theme.DialogContent_borderRadius,
      boxShadow: theme.DialogContent_boxShadow,
      display: 'flex',
      flexDirection: 'column',
      maxHeight: theme.DialogContent_maxHeight,
      minWidth: theme.DialogContent_minWidth,
      pointerEvents: 'all',
      position: 'relative'
    }, getSizeStyles(size));
  },
  ieWrapper: {
    display: 'flex'
  },
  overlay: function overlay(_ref4) {
    var baseTheme = _ref4.theme;

    var theme = componentTheme(baseTheme);

    return {
      backgroundColor: theme.DialogOverlay_backgroundColor,
      bottom: 0,
      left: 0,
      overflow: 'hidden',
      position: 'absolute',
      right: 0,
      top: 0
    };
  },
  root: function root(_ref5) {
    var modeless = _ref5.modeless;
    return {
      alignItems: 'center',
      bottom: 0,
      display: 'flex',
      justifyContent: 'center',
      left: 0,
      position: 'fixed',
      pointerEvents: modeless ? 'none' : undefined,
      right: 0,
      top: 0
    };
  }
};

var Root = (0, _styles.createStyledComponent)('div', styles.root, {
  displayName: 'Dialog',
  filterProps: ['title'],
  includeStyleReset: true
});

var Overlay = (0, _styles.createStyledComponent)('div', styles.overlay, {
  displayName: 'Overlay'
});

var IEWrapper = (0, _styles.createStyledComponent)('div', styles.ieWrapper);

var Content = (0, _styles.createStyledComponent)('div', styles.content, {
  displayName: 'DialogContent'
});

var Animate = (0, _styles.createStyledComponent)('div', styles.animate, {
  displayName: 'Animate'
});

var Animation = (0, _themes.withTheme)(function (_ref6) {
  var children = _ref6.children,
      theme = _ref6.theme,
      restProps = _objectWithoutProperties(_ref6, ['children', 'theme']);

  return _react2.default.createElement(
    _Transition2.default,
    _extends({
      appear: true,
      mountOnEnter: true,
      timeout: parseFloat(componentTheme(theme).Dialog_transitionDuration),
      unmountOnExit: true
    }, restProps),
    function (state) {
      return _jsx(Animate, {
        state: state
      }, void 0, children);
    }
  );
});

var ThemedButton = (0, _themes.createThemedComponent)(_Button2.default, function (_ref7) {
  var theme = _ref7.theme;
  return {
    ButtonIcon_color: theme.color
  };
});

var CloseButton = (0, _styles.createStyledComponent)(ThemedButton, function (_ref8) {
  var baseTheme = _ref8.theme;

  var theme = componentTheme(baseTheme);
  var marginProperty = theme.direction === 'rtl' ? 'marginRight' : 'marginLeft';

  return _defineProperty({}, marginProperty, theme.DialogCloseButton_margin);
}, {
  displayName: 'CloseButton',
  withProps: {
    iconStart: _jsx(_IconClose2.default, {}),
    minimal: true,
    size: 'small'
  }
});

/**
 * Dialog displays content in a layer above the app and requires user
 * interaction to dismiss it. It may appear contextually or as a modal.
 */

var _ref12 = _jsx(Overlay, {});

var Dialog = function (_Component) {
  _inherits(Dialog, _Component);

  function Dialog() {
    var _ref10;

    var _temp, _this, _ret;

    _classCallCheck(this, Dialog);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref10 = Dialog.__proto__ || Object.getPrototypeOf(Dialog)).call.apply(_ref10, [this].concat(args))), _this), _this.state = {
      isExited: false,
      isExiting: false
    }, _this.id = _this.props.id || 'dialog-' + (0, _utils.generateId)(), _this.renderHeader = function () {
      var _this$props = _this.props,
          children = _this$props.children,
          closeButtonLabel = _this$props.closeButtonLabel,
          preventCloseButtonClose = _this$props.preventCloseButtonClose,
          showCloseButton = _this$props.showCloseButton,
          title = _this$props.title,
          variant = _this$props.variant;


      var closeButtonProps = _extends({
        'aria-label': closeButtonLabel
      }, preventCloseButtonClose ? undefined : { onClick: _this.close });

      // prettier-ignore
      var closeButton = showCloseButton ? _react2.default.createElement(CloseButton, closeButtonProps) : undefined;

      var header = void 0;
      var titleProps = {
        id: _this.getTitleId(),
        variant: variant
      };

      if (title) {
        header = _jsx(_DialogHeader2.default, {
          closeButton: closeButton
        }, void 0, typeof title === 'string' ? _react2.default.createElement(
          _DialogTitle2.default,
          titleProps,
          title
        ) : (0, _react.cloneElement)(title, titleProps));
      } else {
        var headerChild = (0, _children.findByType)(children, _DialogHeader2.default);
        if (headerChild) {
          var headerChildProps = {
            titleProps: titleProps,
            closeButton: headerChild.props.closeButton || closeButton
          };
          header = (0, _react.cloneElement)(headerChild, headerChildProps);
        }
      }

      return header;
    }, _this.renderBody = function () {
      var children = _this.props.children;


      var body = void 0;
      var bodyChild = (0, _children.findByType)(children, _DialogBody2.default);
      if (bodyChild) {
        body = bodyChild;
      } else {
        var bodyChildren = (0, _children.excludeByType)(children, [_DialogHeader2.default, _DialogFooter2.default]);
        if (bodyChildren) {
          body = _jsx(_DialogBody2.default, {}, void 0, bodyChildren);
        }
      }

      return body;
    }, _this.renderFooter = function () {
      var _this$props2 = _this.props,
          actions = _this$props2.actions,
          children = _this$props2.children,
          size = _this$props2.size,
          variant = _this$props2.variant;


      var footer = void 0;
      if (actions) {
        var keyedActions = actions.map(function (_ref11, index) {
          var text = _ref11.text,
              buttonSize = _ref11.size,
              restProps = _objectWithoutProperties(_ref11, ['text', 'size']);

          var buttonProps = _extends({
            size: buttonSize || (size === 'large' ? 'large' : 'medium')
          }, restProps);
          return _react2.default.createElement(
            _Button2.default,
            _extends({}, buttonProps, { key: index }),
            text
          );
        });
        footer = _jsx(_DialogFooter2.default, {}, void 0, _jsx(_DialogActions2.default, {
          variant: variant
        }, void 0, keyedActions));
      } else {
        var footerChild = (0, _children.findByType)(children, _DialogFooter2.default);
        if (footerChild) {
          footer = footerChild;
        }
      }

      return footer;
    }, _this.getTitleId = function () {
      return _this.id + '-title';
    }, _this.getContentId = function () {
      return _this.id + '-content';
    }, _this.setContentRef = function (node) {
      _this.dialogContent = node;
    }, _this.setRootRef = function (node) {
      _this.dialogRoot = node;
    }, _this.setLastFocusedElement = function () {
      if (_exenv.canUseDOM) {
        var _global$document = global.document,
            activeElement = _global$document.activeElement,
            body = _global$document.body;

        _this.lastFocusedElement = activeElement && activeElement.focus ? activeElement : body;
      }
    }, _this.restoreFocus = function () {
      _this.lastFocusedElement && _this.lastFocusedElement.focus();
    }, _this.setInitialFocus = function () {
      _this.dialogRoot && _this.dialogRoot.focus();
    }, _this.open = function () {
      var modeless = _this.props.modeless;


      _this.setLastFocusedElement();

      if (!modeless) {
        _this.setAppNode();
        _this.disableAppNode();
        if (_exenv.canUseDOM) {
          _noScroll2.default.on();
        }
      }

      _this.setState({
        isExited: false
      });
    }, _this.close = function () {
      _this.handleExiting();
    }, _this.handleClick = function (event) {
      if (_this.isEventOutsideNode(event, _this.dialogContent)) {
        _this.close();
      }
    }, _this.handleDocumentKeydown = function (event) {
      if (event.key === 'Escape' && !event.defaultPrevented) {
        _this.close();
      }
    }, _this.handleEntered = function () {
      !_this.props.disableFocusOnOpen && _this.setInitialFocus();
      _this.props.onOpen && _this.props.onOpen();
    }, _this.handleExiting = function () {
      _this.setState({
        isExiting: true
      });
    }, _this.handleExited = function () {
      var _this$props3 = _this.props,
          modeless = _this$props3.modeless,
          onClose = _this$props3.onClose;


      _this.setState({
        isExited: true,
        isExiting: false
      }, function () {
        if (!modeless) {
          if (_exenv.canUseDOM) {
            _noScroll2.default.off();
          }
          _this.enableAppNode();
        }

        _this.restoreFocus();
        onClose && onClose();
      });
    }, _this.isEventOutsideNode = function (event, node) {
      var target = event.target;

      return node && target instanceof Node && !node.contains(target);
    }, _this.setAppNode = function () {
      var appSelector = _this.props.appSelector;


      if (appSelector && _exenv.canUseDOM) {
        _this.appNodes = Array.from(document.querySelectorAll(appSelector));

        if (!_this.appNodes.length) {
          throw new Error('[mineral-ui/Dialog]: Unable to find app node(s) using the appSelector of \'' + appSelector + '\'.');
        }
      }
    }, _this.disableAppNode = function () {
      _this.appNodes && _this.appNodes.forEach(function (node) {
        return node.setAttribute('aria-hidden', 'true');
      });
    }, _this.enableAppNode = function () {
      _this.appNodes && _this.appNodes.forEach(function (node) {
        return node.removeAttribute('aria-hidden');
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Dialog, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      if (!prevProps.isOpen && this.props.isOpen) {
        this.open();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          closeOnClickOutside = _props.closeOnClickOutside,
          closeOnEscape = _props.closeOnEscape,
          disableFocusTrap = _props.disableFocusTrap,
          isOpen = _props.isOpen,
          hideOverlay = _props.hideOverlay,
          modeless = _props.modeless,
          size = _props.size,
          title = _props.title,
          usePortal = _props.usePortal,
          restProps = _objectWithoutProperties(_props, ['children', 'closeOnClickOutside', 'closeOnEscape', 'disableFocusTrap', 'isOpen', 'hideOverlay', 'modeless', 'size', 'title', 'usePortal']);

      var _state = this.state,
          isExited = _state.isExited,
          isExiting = _state.isExiting;


      if (isExited) {
        return null;
      }

      var headerId = this.getTitleId();
      var contentId = this.getContentId();
      var hasBody = (0, _children.findByType)(children, _DialogBody2.default);
      var hasImmediatleTitle = (0, _children.findByType)(children, _DialogTitle2.default);
      var hasHeader = title || (0, _children.findByType)(children, _DialogHeader2.default);

      if (hasBody && hasImmediatleTitle) {
        throw new Error('[mineral-ui/Dialog] You must wrap DialogTitle in DialogHeader');
      }
      if (!hasHeader && !this.props['aria-label']) {
        throw new Error('[mineral-ui/Dialog] You must provide an accessible title via the title prop, the DialogTitle component, or the aria-label prop');
      }

      var rootProps = _extends({
        'aria-labelledby': this.props['aria-label'] ? undefined : hasHeader ? headerId : contentId,
        'aria-modal': !modeless,
        id: this.id,
        innerRef: this.setRootRef,
        modeless: modeless,
        role: 'dialog',
        tabIndex: '-1'
      }, closeOnClickOutside && !modeless ? { onClick: this.handleClick } : undefined, restProps);

      var contentProps = {
        id: contentId,
        innerRef: this.setContentRef,
        role: 'document',
        size: size
      };

      var animationProps = {
        in: isOpen && !isExiting,
        onExiting: this.handleExiting,
        onExited: this.handleExited,
        onEntered: this.handleEntered
      };

      var focusTrapProps = {
        active: !disableFocusTrap && !modeless
      };

      var output = _react2.default.createElement(
        Animation,
        animationProps,
        _react2.default.createElement(
          _focusTrapReact2.default,
          focusTrapProps,
          _react2.default.createElement(
            Root,
            rootProps,
            !hideOverlay && !modeless && _ref12,
            _jsx(IEWrapper, {}, void 0, _react2.default.createElement(
              Content,
              contentProps,
              this.renderHeader(),
              this.renderBody(),
              this.renderFooter()
            )),
            closeOnEscape && _jsx(_EventListener2.default, {
              listeners: [{
                target: 'document',
                event: 'keydown',
                handler: this.handleDocumentKeydown
              }]
            })
          )
        )
      );

      return usePortal ? _jsx(_Portal2.default, {}, void 0, output) : output;
    }
  }]);

  return Dialog;
}(_react.Component);

Dialog.defaultProps = {
  closeButtonLabel: 'close',
  closeOnClickOutside: true,
  closeOnEscape: true,
  size: 'medium',
  usePortal: true
};
exports.default = Dialog;
Dialog.propTypes = process.env.NODE_ENV !== "production" ? {
  /**
   * Configuration for the [Buttons](/components/button) rendered at the bottom
   * of Dialog; accepts all Button props
   */
  actions: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired,
    size: _propTypes2.default.oneOf(['small', 'medium', 'large', 'jumbo'])
  }).isRequired),

  /**
   * CSS selector matching the node(s) which will be accessibly hidden (unless
   * \`modeless\`)
   */
  appSelector: _propTypes2.default.string,

  /** @Private Accessible label */
  "aria-label": _propTypes2.default.string,

  /**
   * Content of Dialog (see the [Basic](#basic) and
   * [Alternate Syntax](#alternate) examples)
   */
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node) : _propTypes2.default.any).apply(this, arguments);
  },

  /** Accessible label for the close button */
  closeButtonLabel: _propTypes2.default.string,

  /** Close Dialog with the 'Escape' key */
  closeOnEscape: _propTypes2.default.bool,

  /** Close Dialog by clicking outside of its content */
  closeOnClickOutside: _propTypes2.default.bool,

  /** Disable focus being placed on Dialog upon opening */
  disableFocusOnOpen: _propTypes2.default.bool,

  /** Disable focus being trapped within the Dialog when open */
  disableFocusTrap: _propTypes2.default.bool,

  /** Hide the overlay underneath Dialog */
  hideOverlay: _propTypes2.default.bool,

  /** Id of the Dialog */
  id: _propTypes2.default.string,

  /** Dialog only renders its content when true */
  isOpen: _propTypes2.default.bool,

  /** Renders Dialog without modal behavior ([see example](#modeless)) */
  modeless: _propTypes2.default.bool,

  /** Called when Dialog is closed */
  onClose: _propTypes2.default.func,

  /** Called when Dialog is opened */
  onOpen: _propTypes2.default.func,

  /** @Private Prevent the dialog from closing when the close button is clicked */
  preventCloseButtonClose: _propTypes2.default.bool,

  /** Render a close button for Dialog */
  showCloseButton: _propTypes2.default.bool,

  /** Available sizes */
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),

  /**
   * Title of the dialog; must be a string or
   * [DialogTitle](/components/dialog-title)
   */
  title: _propTypes2.default.oneOfType([_propTypes2.default.string, function () {
    return (typeof React$Element === 'function' ? _propTypes2.default.instanceOf(React$Element) : _propTypes2.default.any).apply(this, arguments);
  }]),

  /** @Private Use a Portal to render Dialog to the body */
  usePortal: _propTypes2.default.bool,

  /** Available variants */
  variant: _propTypes2.default.oneOf(['danger', 'success', 'warning'])
} : {};
