// mineral-ui/Dialog/DialogActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = DialogActions;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Button = require('../Button');

var _Button2 = _interopRequireDefault(_Button);

var _children = require('../utils/children');

var _styles = require('../styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    DialogActionsItem_margin: baseTheme.space_stack_sm

  }, baseTheme);
};

var Root = (0, _styles.createStyledComponent)('div', function (_ref) {
  var baseTheme = _ref.theme;

  var theme = componentTheme(baseTheme);
  var marginProperty = theme.direction === 'rtl' ? 'marginLeft' : 'marginRight';

  return {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'flex-end',

    '& > *:not(:last-child)': _defineProperty({}, marginProperty, theme.DialogActionsItem_margin)
  };
}, {
  displayName: 'DialogActions'
});

/**
 * DialogActions renders primary and secondary actions inside
 * [Dialog](/components/dialog).
 */
function DialogActions(props) {
  var children = props.children,
      variant = props.variant,
      rootProps = _objectWithoutProperties(props, ['children', 'variant']);

  var actions = (0, _children.findAllByType)(children, _Button2.default);
  var lastIndex = actions.length - 1;
  actions = actions.map(function (action, index) {
    var props = action.props;
    /* eslint-disable react/prop-types */

    return (0, _react.cloneElement)(action, {
      key: index,
      minimal: props.minimal !== undefined ? props.minimal : index !== lastIndex,
      primary: props.primary !== undefined ? props.primary : index === lastIndex,
      variant: props.variant !== undefined ? props.variant : index === lastIndex ? variant : undefined
    });
    /* eslint-enable react/prop-types */
  });

  return _react2.default.createElement(
    Root,
    rootProps,
    actions
  );
}
DialogActions.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Rendered [Button(s)](/components/button) */
  children: _propTypes2.default.oneOfType([function () {
    return (typeof React$Element === 'function' ? _propTypes2.default.instanceOf(React$Element) : _propTypes2.default.any).apply(this, arguments);
  }, _propTypes2.default.arrayOf(function () {
    return (typeof React$Element === 'function' ? _propTypes2.default.instanceOf(React$Element).isRequired : _propTypes2.default.any.isRequired).apply(this, arguments);
  })]),

  /** Available variants */
  variant: _propTypes2.default.oneOf(['danger', 'success', 'warning'])
} : {};
