// mineral-ui/es/Card/CardTitleMenu.js
var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import { createStyledComponent, pxToEm } from '../styles';
import Button from '../Button';
import Dropdown from '../Dropdown';
import IconMoreHoriz from '../Icon/IconMoreHoriz';

/*
 * A large Button, even with zero'd padding, is still a bit too large in this
 * context. These styles allow the Button to shrink, but the Icon remains the
 * same size.
 */
var MenuButton = createStyledComponent(Button, function (_ref) {
  var theme = _ref.theme;
  return {
    height: 'auto',
    minWidth: 0,
    overflow: 'hidden',
    padding: 0,

    // Inner
    '& > span': {
      display: 'block',
      margin: '-' + pxToEm(2)
    },

    // Icon
    '& [role="img"]': {
      color: theme.icon_color
    }
  };
});

/**
 * CardTitleMenu
 */
var CardTitleMenu = function CardTitleMenu(props) {
  var triggerTitle = props.triggerTitle,
      restProps = _objectWithoutProperties(props, ['triggerTitle']);

  var rootProps = _extends({
    placement: 'bottom-end'
  }, restProps);
  var buttonProps = {
    iconStart: _jsx(IconMoreHoriz, {
      title: triggerTitle
    }),
    minimal: true
  };

  return React.createElement(
    Dropdown,
    rootProps,
    React.createElement(MenuButton, buttonProps)
  );
};

CardTitleMenu.propTypes = process.env.NODE_ENV !== "production" ? {
  /** See the [Actions Menu](#actions-menu) example */
  data: PropTypes.oneOfType([function () {
    return (typeof bpfrpt_proptype_Items === 'function' ? bpfrpt_proptype_Items : PropTypes.shape(bpfrpt_proptype_Items)).apply(this, arguments);
  }, function () {
    return (typeof bpfrpt_proptype_ItemGroups === 'function' ? bpfrpt_proptype_ItemGroups : PropTypes.shape(bpfrpt_proptype_ItemGroups)).apply(this, arguments);
  }]),

  /** Title for the actions menu [Icon](/components/Icon) */
  triggerTitle: PropTypes.string
} : {};
CardTitleMenu.defaultProps = {
  triggerTitle: 'Card actions'
};

export default CardTitleMenu;
import { bpfrpt_proptype_Items } from '../Menu/Menu';
import { bpfrpt_proptype_ItemGroups } from '../Menu/Menu';
import PropTypes from 'prop-types';
