// mineral-ui/Link/Link.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _memoizeOne = require('memoize-one');

var _memoizeOne2 = _interopRequireDefault(_memoizeOne);

var _styles = require('../styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    Link_borderColor_focus: baseTheme.borderColor_theme_focus,
    Link_color: baseTheme.color_theme,
    Link_color_active: baseTheme.color_theme_active,
    Link_color_hover: baseTheme.color_theme_hover,
    Link_color_focus: baseTheme.color_theme_focus

  }, baseTheme);
};

var styles = function styles(_ref) {
  var variant = _ref.variant,
      baseTheme = _ref.theme;

  var theme = componentTheme(baseTheme);

  if (variant) {
    // prettier-ignore
    theme = _extends({}, theme, {
      Link_borderColor_focus: baseTheme['borderColor_' + variant + '_focus'],
      Link_color: baseTheme['color_' + variant],
      Link_color_hover: baseTheme['color_' + variant + '_hover'],
      Link_color_focus: baseTheme['color_' + variant + '_focus'],
      Link_color_active: baseTheme['color_' + variant + '_active']
    });
  }

  return {
    color: theme.Link_color,
    textDecoration: 'none',

    '&:hover': {
      color: theme.Link_color_hover,
      textDecoration: 'underline'
    },
    '&:focus': {
      color: theme.Link_color_focus,
      outline: '1px solid ' + theme.Link_borderColor_focus,
      outlineOffset: '2px'
    },
    // `:active` must be last, to follow LVHFA order:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
    '&:active': {
      color: theme.Link_color_active
    }
  };
};

var createRootNode = function createRootNode(props) {
  var _props$element = props.element,
      element = _props$element === undefined ? Link.defaultProps.element : _props$element;


  return (0, _styles.createStyledComponent)(element, styles, {
    displayName: 'Link',
    filterProps: ['element', 'variant'],
    rootEl: element
  });
};

/**
 * The Link component creates a hyperlink to external pages, files, anchors on
 * the same page, or another URL.
 */

var Link = function (_Component) {
  _inherits(Link, _Component);

  function Link() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, Link);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = Link.__proto__ || Object.getPrototypeOf(Link)).call.apply(_ref2, [this].concat(args))), _this), _this.getRootNode = (0, _memoizeOne2.default)(createRootNode, function (nextProps, prevProps) {
      return nextProps.element === prevProps.element;
    }), _temp), _possibleConstructorReturn(_this, _ret);
  }

  // Must be an instance method to avoid affecting other instances memoized keys


  _createClass(Link, [{
    key: 'render',
    value: function render() {
      var Root = this.getRootNode(this.props);

      return _react2.default.createElement(Root, this.props);
    }
  }]);

  return Link;
}(_react.Component);

Link.defaultProps = {
  element: 'a'
};
exports.default = Link;
Link.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Content of the Link */
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node) : _propTypes2.default.any).apply(this, arguments);
  },

  /** A URL or URL fragment to which the Link points */
  href: _propTypes2.default.string,

  /** Element to be used as the root node - e.g. `a` or `ReactRouterLink` */
  element: function element() {
    return (typeof $FlowFixMe === 'function' ? _propTypes2.default.instanceOf($FlowFixMe) : _propTypes2.default.any).apply(this, arguments);
  },
  // Should allow string | React class
  /** Available variants */
  variant: _propTypes2.default.oneOf(['danger', 'success', 'warning'])
} : {};
