// mineral-ui/Card/CardImage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CardImage;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('../styles');

var _Card = require('./Card');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Root = (0, _styles.createStyledComponent)('img', function (props) {
  var theme = (0, _Card.componentTheme)(props.theme);

  return {
    display: 'block',
    marginBottom: theme.CardRow_marginVertical,
    marginTop: theme.CardRow_marginVertical,
    maxWidth: '100%',

    '&:first-child': {
      borderRadius: theme.Card_borderRadius + ' ' + theme.Card_borderRadius + ' 0 0',
      marginTop: 0
    },

    '&:last-child': {
      borderRadius: '0 0 ' + theme.Card_borderRadius + ' ' + theme.Card_borderRadius,
      marginBottom: 0
    }
  };
}, {
  displayName: 'CardImage',
  rootEl: 'img'
});

/**
 * CardImage renders images full-bleed inside of a [Card](/components/card).
 * Use CardImage to display static media.
 */

function CardImage(props) {
  return _react2.default.createElement(Root, props);
}
