// mineral-ui/es/Table/TableBase.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import createReactContext from 'create-react-context';
import { createStyledComponent } from '../styles';
import { createThemedComponent, mapComponentThemes } from '../themes';
import { generateId } from '../utils';
import _OverflowContainer, { componentTheme as overflowContainerComponentTheme } from '../OverflowContainer/OverflowContainer';
import TableDataRow from './TableDataRow';
import TableHeaderRow from './TableHeaderRow';
import TableBody from './TableBody';
import TableHeader from './TableHeader';
import TableTitle from './TableTitle';

// See Table


export var componentTheme = function componentTheme(baseTheme) {
  return mapComponentThemes({
    name: 'OverflowContainer',
    theme: overflowContainerComponentTheme(baseTheme)
  }, {
    name: 'Table',
    theme: {}
  }, baseTheme);
};

var OverflowContainer = createThemedComponent(_OverflowContainer, function (_ref) {
  var baseTheme = _ref.theme;
  return mapComponentThemes({
    name: 'Table',
    theme: componentTheme(baseTheme)
  }, {
    name: 'OverflowContainer',
    theme: {}
  }, baseTheme);
});

var Root = createStyledComponent('table', {
  borderCollapse: 'collapse',
  borderSpacing: 0,
  width: '100%'
}, {
  displayName: 'Table',
  rootEl: 'table',
  includeStyleReset: true
});

export var TableContext = createReactContext({});

/**
 * Table displays structured data with columns and rows.
 */

var TableBase = function (_Component) {
  _inherits(TableBase, _Component);

  function TableBase() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, TableBase);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = TableBase.__proto__ || Object.getPrototypeOf(TableBase)).call.apply(_ref2, [this].concat(args))), _this), _this.id = _this.props.id || 'table-' + generateId(), _this.titleId = _this.id + '-title', _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TableBase, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          columns = _props.columns,
          data = _props.data,
          density = _props.density,
          hideHeader = _props.hideHeader,
          hideTitle = _props.hideTitle,
          highContrast = _props.highContrast,
          isSortable = _props.isSortable,
          messages = _props.messages,
          rowKey = _props.rowKey,
          scrollable = _props.scrollable,
          selectable = _props.selectable,
          sortable = _props.sortable,
          striped = _props.striped,
          title = _props.title,
          titleAppearance = _props.titleAppearance,
          titleElement = _props.titleElement,
          rootProps = _objectWithoutProperties(_props, ['columns', 'data', 'density', 'hideHeader', 'hideTitle', 'highContrast', 'isSortable', 'messages', 'rowKey', 'scrollable', 'selectable', 'sortable', 'striped', 'title', 'titleAppearance', 'titleElement']);

      var appearanceProps = {
        density: density,
        highContrast: highContrast,
        striped: striped
      };
      var tableHeaderRowProps = {
        checked: selectable && selectable.all,
        columns: columns,
        indeterminate: selectable && selectable.some,
        isSortable: isSortable,
        messages: messages,
        sortable: sortable,
        toggle: selectable && selectable.toggleAll
      };

      var table = _jsx(TableContext.Provider, {
        value: appearanceProps
      }, void 0, React.createElement(
        Root,
        rootProps,
        _jsx(TableTitle, {
          appearance: titleAppearance,
          element: titleElement,
          hide: hideTitle,
          id: this.titleId
        }, void 0, title),
        _jsx(TableHeader, {
          hide: hideHeader
        }, void 0, React.createElement(TableHeaderRow, tableHeaderRowProps)),
        _jsx(TableBody, {}, void 0, data.map(function (rowData, index) {
          var rowProps = {
            checked: selectable && selectable.isSelected(rowData),
            columns: columns,
            data: rowData,
            messages: messages,
            rowIndex: index,
            toggle: selectable && selectable.toggle
          };
          return React.createElement(TableDataRow, _extends({ key: rowData[rowKey] || index }, rowProps));
        }))
      ));

      if (scrollable) {
        var containerProps = {
          'aria-labelledby': this.titleId,
          role: 'group',
          scrollX: true
        };
        table = React.createElement(
          OverflowContainer,
          containerProps,
          table
        );
      }

      return table;
    }
  }]);

  return TableBase;
}(Component);

TableBase.defaultProps = {
  titleElement: 'h4',
  density: 'compact',
  scrollable: true
};
export default TableBase;
TableBase.propTypes = process.env.NODE_ENV !== "production" ? {
  columns: function columns() {
    return (typeof bpfrpt_proptype_Columns === 'function' ? bpfrpt_proptype_Columns.isRequired ? bpfrpt_proptype_Columns.isRequired : bpfrpt_proptype_Columns : PropTypes.shape(bpfrpt_proptype_Columns).isRequired).apply(this, arguments);
  },
  data: function data() {
    return (typeof bpfrpt_proptype_Rows === 'function' ? bpfrpt_proptype_Rows.isRequired ? bpfrpt_proptype_Rows.isRequired : bpfrpt_proptype_Rows : PropTypes.shape(bpfrpt_proptype_Rows).isRequired).apply(this, arguments);
  },
  hideHeader: PropTypes.bool,
  hideTitle: PropTypes.bool,
  id: PropTypes.string,
  isSortable: PropTypes.bool,
  messages: function messages() {
    return (typeof bpfrpt_proptype_Messages === 'function' ? bpfrpt_proptype_Messages.isRequired ? bpfrpt_proptype_Messages.isRequired : bpfrpt_proptype_Messages : PropTypes.shape(bpfrpt_proptype_Messages).isRequired).apply(this, arguments);
  },
  rowKey: PropTypes.string,
  scrollable: PropTypes.bool,
  selectable: function selectable() {
    return (typeof bpfrpt_proptype_SelectableType === 'function' ? bpfrpt_proptype_SelectableType : PropTypes.shape(bpfrpt_proptype_SelectableType)).apply(this, arguments);
  },
  sortable: function sortable() {
    return (typeof bpfrpt_proptype_SortableType === 'function' ? bpfrpt_proptype_SortableType : PropTypes.shape(bpfrpt_proptype_SortableType)).apply(this, arguments);
  },
  title: function title() {
    return (typeof React$Node === 'function' ? PropTypes.instanceOf(React$Node).isRequired : PropTypes.any.isRequired).apply(this, arguments);
  },
  titleAppearance: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  titleElement: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])
} : {};
import { bpfrpt_proptype_SelectableType } from './Selectable';
import { bpfrpt_proptype_SortableType } from './Sortable';
import { bpfrpt_proptype_Columns } from './Table';
import { bpfrpt_proptype_Messages } from './Table';
import { bpfrpt_proptype_Rows } from './Table';
import PropTypes from 'prop-types';
