// mineral-ui/es/Table/Table.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import deepEqual from 'react-fast-compare';
import Selectable from './Selectable';
import Sortable from './Sortable';
import TableBase from './TableBase';

var bpfrpt_proptype_Columns = PropTypes.arrayOf(function () {
  return (typeof Column === 'function' ? PropTypes.instanceOf(Column).isRequired : PropTypes.any.isRequired).apply(this, arguments);
});
var bpfrpt_proptype_Column = {
  cell: function cell() {
    return (typeof RenderFn === 'function' ? PropTypes.instanceOf(RenderFn) : PropTypes.any).apply(this, arguments);
  },
  content: function content() {
    return (typeof React$Node === 'function' ? PropTypes.instanceOf(React$Node).isRequired : PropTypes.any.isRequired).apply(this, arguments);
  },
  headerCell: function headerCell() {
    return (typeof RenderFn === 'function' ? PropTypes.instanceOf(RenderFn) : PropTypes.any).apply(this, arguments);
  },
  key: PropTypes.string.isRequired,
  label: PropTypes.string,
  maxWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  minWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  primary: PropTypes.bool,
  sortable: PropTypes.bool,
  sortComparator: PropTypes.func,
  textAlign: PropTypes.oneOf(['start', 'end', 'center', 'justify']),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};

// See demos/Table/index.js additionalTypes for descriptions

var bpfrpt_proptype_Messages = {
  deselectAllRows: PropTypes.string.isRequired,
  deselectRow: PropTypes.string.isRequired,
  selectAllRows: PropTypes.string.isRequired,
  selectRow: PropTypes.string.isRequired,
  selectedRows: PropTypes.string.isRequired,
  sortColumnAscending: PropTypes.string.isRequired,
  sortColumnDescending: PropTypes.string.isRequired
};
var bpfrpt_proptype_Row = PropTypes.object;
var bpfrpt_proptype_Rows = PropTypes.arrayOf(PropTypes.object.isRequired);
var bpfrpt_proptype_RenderFn = PropTypes.func;
var bpfrpt_proptype_RenderProps = {
  props: PropTypes.object.isRequired,
  state: PropTypes.object,
  helpers: PropTypes.object
};


var generateColumns = function generateColumns(data) {
  return data[0] ? Object.keys(data[0]).reduce(function (acc, value) {
    acc.push({ content: value, key: value });
    return acc;
  }, []) : [];
};

var getColumns = function getColumns(_ref) {
  var columns = _ref.columns,
      data = _ref.data;
  return columns || generateColumns(data);
};

var getComparators = function getComparators(_ref2) {
  var columns = _ref2.columns;

  var comparators = columns && columns.reduce(function (acc, column) {
    var key = column.key,
        sortComparator = column.sortComparator;

    if (sortComparator) {
      acc[key] = sortComparator;
    }
    return acc;
  }, {});

  return comparators && Object.keys(comparators).length ? comparators : undefined;
};

var getSelectableRows = function getSelectableRows(rows) {
  return rows.filter(function (row) {
    return !row.disabled;
  });
};

var getSortable = function getSortable(_ref3) {
  var columns = _ref3.columns,
      defaultSort = _ref3.defaultSort,
      sort = _ref3.sort,
      sortable = _ref3.sortable;
  return Boolean(defaultSort || sort || sortable || columns && columns.some(function (column) {
    return column.sortable;
  }));
};

/**
 * Table displays structured data with sortable columns and selectable rows.
 */

var Table = function (_Component) {
  _inherits(Table, _Component);

  function Table() {
    var _ref4;

    var _temp, _this, _ret;

    _classCallCheck(this, Table);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref4 = Table.__proto__ || Object.getPrototypeOf(Table)).call.apply(_ref4, [this].concat(args))), _this), _this.columns = getColumns(_this.props), _this.comparators = getComparators(_this.props), _this.selectableRows = getSelectableRows(_this.props.data), _this.sortable = getSortable(_this.props), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Table, [{
    key: 'componentWillUpdate',
    value: function componentWillUpdate(nextProps) {
      var columnsChanged = !deepEqual(this.props.columns, nextProps.columns);
      var dataChanged = !deepEqual(this.props.data, nextProps.data);

      if (columnsChanged || !this.props.columns && dataChanged) {
        this.columns = getColumns(nextProps);
      }

      if (columnsChanged) {
        this.sortable = getSortable(nextProps);
        this.comparators = getComparators(nextProps);
      }

      if (dataChanged) {
        this.selectableRows = getSelectableRows(nextProps.data);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          defaultSelectedRows = _props.defaultSelectedRows,
          onToggleRow = _props.onToggleRow,
          onToggleAllRows = _props.onToggleAllRows,
          selectable = _props.selectable,
          selectedRows = _props.selectedRows,
          sortable = _props.sortable,
          restProps = _objectWithoutProperties(_props, ['defaultSelectedRows', 'onToggleRow', 'onToggleAllRows', 'selectable', 'selectedRows', 'sortable']);

      var rootProps = _extends({}, restProps, {
        columns: this.columns,
        comparators: this.comparators
      }, defaultSelectedRows ? { defaultSelected: getSelectableRows(defaultSelectedRows) } : undefined, onToggleRow ? { onToggle: onToggleRow } : undefined, onToggleAllRows ? { onToggleAll: onToggleAllRows } : undefined, selectedRows ? { selected: selectedRows } : undefined, selectable ? { selectableRows: this.selectableRows } : undefined, this.sortable ? { sortable: sortable } : undefined);

      if (selectable && this.sortable) {
        return React.createElement(SelectableSortableTable, rootProps);
      } else if (selectable) {
        return React.createElement(SelectableTable, rootProps);
      } else if (this.sortable) {
        return React.createElement(SortableTable, rootProps);
      } else {
        var ignoreSortable = rootProps.sortable,
            rootPropsWithoutSortable = _objectWithoutProperties(rootProps, ['sortable']);

        return React.createElement(TableBase, rootPropsWithoutSortable);
      }
    }
  }]);

  return Table;
}(Component);

/* eslint-disable react/prop-types */


Table.defaultProps = {
  density: 'compact',
  messages: {
    deselectAllRows: 'Deselect all rows',
    deselectRow: 'Deselect row',
    selectAllRows: 'Select all rows',
    selectedRows: 'Selected rows',
    selectRow: 'Select row',
    sortColumnAscending: 'Sort column in ascending order',
    sortColumnDescending: 'Sort column in descending order'
  },
  titleElement: 'h4'
};
Table.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Column definitions ([see Column type for details](#Column-type)) */
  columns: function columns() {
    return (typeof Columns === 'function' ? PropTypes.instanceOf(Columns) : PropTypes.any).apply(this, arguments);
  },

  /** Row data ([see example for more details](#basic)) */
  data: function data() {
    return (typeof Rows === 'function' ? PropTypes.instanceOf(Rows).isRequired : PropTypes.any.isRequired).apply(this, arguments);
  },

  /** Initially selected rows when `selectable = true`. Primarily for use with uncontrolled components. */
  defaultSelectedRows: PropTypes.arrayOf(PropTypes.object.isRequired),

  /** Initially sorted column & direction. Primarily for use with uncontrolled components. */
  defaultSort: PropTypes.shape({
    key: PropTypes.string.isRequired,
    descending: PropTypes.bool
  }),

  /** Amount of vertical space in Table's cells */
  density: PropTypes.oneOf(['compact', 'spacious']),

  /** Visually hide Table's header, but keep available for [assistive technologies](https://webaccess.berkeley.edu/resources/assistive-technology) */
  hideHeader: PropTypes.bool,

  /** Visually hide Table's title, but keep available for [assistive technologies](https://webaccess.berkeley.edu/resources/assistive-technology) */
  hideTitle: PropTypes.bool,

  /** Render Table with high-contrast styles */
  highContrast: PropTypes.bool,

  /**
   * Various messages and labels used by Table
   * ([see example for more details](#rtl))
   */
  messages: function messages() {
    return (typeof Messages === 'function' ? PropTypes.instanceOf(Messages) : PropTypes.any).apply(this, arguments);
  },

  /** Called when data is sorted */
  onSort: PropTypes.func,

  /** Called when all rows are selected/deselected */
  onToggleAllRows: PropTypes.func,

  /** Called when a single row is selected/deselected */
  onToggleRow: PropTypes.func,

  /**
   * Specifies a key in the row data that gives a row its unique identity.
   * See the [React docs](https://reactjs.org/docs/lists-and-keys.html#keys).
   */
  rowKey: PropTypes.string,

  /**
   * Determines the scrolling behavior when Table's width exceeds that of its
   * container
   */
  scrollable: PropTypes.bool,

  /**
   * Enable the user to select rows. Prepends a column for checkboxes to your
   * Table.
   */
  selectable: PropTypes.bool,

  /**
   * Selected rows when `selectable = true`. Primarily for use with controlled
   * components.
   */
  selectedRows: PropTypes.arrayOf(PropTypes.object.isRequired),

  /**
   * Sorted column & direction
   */
  sort: PropTypes.shape({
    key: PropTypes.string.isRequired,
    descending: PropTypes.bool
  }),

  /** Enable the user to sort all columns */
  sortable: PropTypes.bool,

  /**
   * The [sort comparator function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description)
   * used by sortable columns
   */
  sortComparator: PropTypes.func,

  /** Renders Table with alternating row stripes */
  striped: PropTypes.bool,

  /** Title for Table */
  title: function title() {
    return (typeof React$Node === 'function' ? PropTypes.instanceOf(React$Node).isRequired : PropTypes.any.isRequired).apply(this, arguments);
  },

  /** Available title styles (see [Text](/components/text)) */
  titleAppearance: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),

  /** Available title elements (see [Text](/components/text)) */
  titleElement: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])
} : {};
var SelectableTable = function SelectableTable(props) {
  var data = props.data,
      selectableRows = props.selectableRows;

  return React.createElement(
    Selectable,
    _extends({}, props, { data: selectableRows }),
    function (props) {
      return React.createElement(TableBase, _extends({}, props, { data: data }));
    }
  );
};

var SortableTable = function SortableTable(props) {
  return React.createElement(
    Sortable,
    _extends({}, props, { isSortable: props.sortable }),
    function (_ref5) {
      var props = _objectWithoutProperties(_ref5, []);

      return React.createElement(TableBase, _extends({}, props, { data: props.sortable.data }));
    }
  );
};

var SelectableSortableTable = function SelectableSortableTable(props) {
  var data = props.data,
      selectableRows = props.selectableRows;

  return React.createElement(
    Selectable,
    _extends({}, props, { data: selectableRows }),
    function (props) {
      return React.createElement(SortableTable, _extends({}, props, { data: data }));
    }
  );
};
/* eslint-enable */

export default Table;
import { bpfrpt_proptype_Comparators } from './Sortable';
import PropTypes from 'prop-types';
