// mineral-ui/Table/TableTitle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('../styles');

var _themes = require('../themes');

var _Text = require('../Text');

var _Text2 = _interopRequireDefault(_Text);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var bpfrpt_proptype_TitleAppearance = _propTypes2.default.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    TableTitle_color: baseTheme.h4_color,
    TableTitle_fontSize: baseTheme.h4_fontSize,
    TableTitle_fontWeight: baseTheme.h4_fontWeight,
    TableTitle_marginBottom: baseTheme.space_stack_sm

  }, baseTheme);
};

var Root = (0, _styles.createStyledComponent)('caption', function (_ref) {
  var hide = _ref.hide,
      baseTheme = _ref.theme;

  var theme = componentTheme(baseTheme);

  return _extends({
    marginBottom: theme.TableTitle_marginBottom
  }, hide ? {
    border: '0',
    clip: 'rect(0 0 0 0)',
    clipPath: 'inset(50%)',
    height: '1px',
    margin: '-1px',
    overflow: 'hidden',
    padding: '0',
    position: 'absolute',
    whiteSpace: 'nowrap',
    width: '1px'
  } : undefined);
}, {
  displayName: 'Caption',
  rootEl: 'caption'
});

/**
 * TableTitle
 */
var TableTitle = function TableTitle(_ref2) {
  var hide = _ref2.hide,
      id = _ref2.id,
      theme = _ref2.theme,
      restProps = _objectWithoutProperties(_ref2, ['hide', 'id', 'theme']);

  var rootProps = _extends({
    hide: hide
  }, restProps);

  var TitleContent = function TitleContent(props) {
    return _react2.default.createElement(_Text2.default, _extends({ align: 'start', id: id }, props));
  };

  var title = (0, _Text.textWithThemeOverrides)(_extends({
    displayName: 'TableTitle',
    textComponent: TitleContent,
    theme: theme
  }, restProps));

  return _react2.default.createElement(
    Root,
    rootProps,
    title
  );
};

TableTitle.propTypes = process.env.NODE_ENV !== "production" ? {
  appearance: function appearance() {
    return (typeof TitleAppearance === 'function' ? _propTypes2.default.instanceOf(TitleAppearance) : _propTypes2.default.any).apply(this, arguments);
  },
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node).isRequired : _propTypes2.default.any.isRequired).apply(this, arguments);
  },
  element: function element() {
    return (typeof TitleAppearance === 'function' ? _propTypes2.default.instanceOf(TitleAppearance) : _propTypes2.default.any).apply(this, arguments);
  },
  hide: _propTypes2.default.bool,
  id: _propTypes2.default.string.isRequired,
  theme: _propTypes2.default.object.isRequired
} : {};
exports.default = (0, _themes.withTheme)(TableTitle);
