// mineral-ui/Table/TableHeaderCell.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ThemedTableCell = exports.componentTheme = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _memoizeOne = require('memoize-one');

var _memoizeOne2 = _interopRequireDefault(_memoizeOne);

var _utils = require('../utils');

var _styles = require('../styles');

var _themes = require('../themes');

var _TableCell = require('./TableCell');

var _TableCell2 = _interopRequireDefault(_TableCell);

var _TableBase = require('./TableBase');

var _Table = require('./Table');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// prettier-ignore
var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return (0, _themes.mapComponentThemes)({
    name: 'TableCell',
    theme: (0, _TableCell.componentTheme)(baseTheme)
  }, {
    name: 'TableHeaderCell',
    theme: {
      TableHeaderCell_borderVertical: '1px dotted ' + baseTheme.borderColor,
      TableHeaderCell_borderVertical_highContrast: '1px dotted ' + baseTheme.color_gray_80,
      TableHeaderCell_fontWeight: baseTheme.fontWeight_bold,
      TableHeaderCell_paddingHorizontal: baseTheme.space_inline_md,
      TableHeaderCell_paddingVertical: (0, _styles.pxToEm)(12),
      TableHeaderCell_paddingVertical_spacious: baseTheme.space_stack_md,
      TableHeaderCell_verticalAlign: 'bottom'
    }
  }, baseTheme);
};

var ThemedTableCell = exports.ThemedTableCell = (0, _themes.createThemedComponent)(_TableCell2.default, function (_ref) {
  var baseTheme = _ref.theme;
  return (0, _themes.mapComponentThemes)({
    name: 'TableHeaderCell',
    theme: componentTheme(baseTheme)
  }, {
    name: 'TableCell',
    theme: {}
  }, baseTheme, ['TableHeaderCell_borderVertical', 'TableHeaderCell_fontSize', 'TableHeaderCell_paddingHorizontal', 'TableHeaderCell_paddingVertical', 'TableHeaderCell_verticalAlign']);
});

var REGEX_IS_EM_VALUE = /\d+em$/;
var getWidth = function getWidth(value, fontSize) {
  return REGEX_IS_EM_VALUE.test(value) ? (0, _styles.getNormalizedValue)(value, fontSize) : value;
};

var styles = function styles(_ref2) {
  var _before, _notFirstChild;

  var highContrast = _ref2.highContrast,
      maxWidth = _ref2.maxWidth,
      minWidth = _ref2.minWidth,
      baseTheme = _ref2.theme,
      width = _ref2.width;

  var theme = componentTheme(baseTheme);
  var fontSize = theme.TableHeaderCell_fontSize;
  var rtl = theme.direction === 'rtl';
  var borderProperty = rtl ? 'borderRight' : 'borderLeft';
  var borderVertical = highContrast ? theme.TableHeaderCell_borderVertical_highContrast : theme.TableHeaderCell_borderVertical;
  var positionProperty = rtl ? 'right' : 'left';

  return {
    fontWeight: theme.TableHeaderCell_fontWeight,
    maxWidth: getWidth(maxWidth, fontSize),
    minWidth: getWidth(minWidth, fontSize),
    position: 'relative',
    width: getWidth(width, fontSize),
    zIndex: 1,

    // Using this "border" to appease Firefox, which extends TableHeaderCell's
    // real border down the entire column after clicking a TableSortableHeaderCell.
    '&:not(:first-child)': (_notFirstChild = {}, _defineProperty(_notFirstChild, borderProperty, 0), _defineProperty(_notFirstChild, '&::before', (_before = {}, _defineProperty(_before, borderProperty, borderVertical), _defineProperty(_before, 'bottom', 0), _defineProperty(_before, 'content', '""'), _defineProperty(_before, positionProperty, 0), _defineProperty(_before, 'position', 'absolute'), _defineProperty(_before, 'top', 0), _defineProperty(_before, 'width', 0), _defineProperty(_before, 'zIndex', -1), _before)), _notFirstChild)
  };
};

var createRootNode = function createRootNode(props) {
  var _props$element = props.element,
      element = _props$element === undefined ? TableHeaderCell.defaultProps.element : _props$element;


  return (0, _styles.createStyledComponent)(ThemedTableCell, styles, {
    displayName: 'TableHeaderCell',
    filterProps: ['width'],
    forwardProps: ['element', 'noPadding', 'textAlign'],
    rootEl: element,
    withProps: { element: element }
  });
};

/**
 * TableHeaderCell
 */

var TableHeaderCell = function (_PureComponent) {
  _inherits(TableHeaderCell, _PureComponent);

  function TableHeaderCell() {
    var _ref3;

    var _temp, _this, _ret;

    _classCallCheck(this, TableHeaderCell);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref3 = TableHeaderCell.__proto__ || Object.getPrototypeOf(TableHeaderCell)).call.apply(_ref3, [this].concat(args))), _this), _this.getRootNode = (0, _memoizeOne2.default)(createRootNode, function (nextProps, prevProps) {
      return nextProps.element === prevProps.element;
    }), _temp), _possibleConstructorReturn(_this, _ret);
  }

  // Must be an instance method to avoid affecting other instances memoized keys


  _createClass(TableHeaderCell, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _jsx(_TableBase.TableContext.Consumer, {}, void 0, function (tableContextProps) {
        var _props = _this2.props,
            label = _props.label,
            render = _props.render,
            restProps = _objectWithoutProperties(_props, ['label', 'render']);

        var rootProps = _extends({}, tableContextProps, {
          'aria-label': label
        }, restProps);

        if ((0, _utils.isRenderProp)(render)) {
          return render({
            props: _extends({}, rootProps, {
              label: label
            })
          });
        }

        var Root = _this2.getRootNode(_this2.props);
        return _react2.default.createElement(Root, rootProps);
      });
    }
  }]);

  return TableHeaderCell;
}(_react.PureComponent);

TableHeaderCell.defaultProps = {
  element: 'th',
  textAlign: 'start'
};
exports.default = TableHeaderCell;
TableHeaderCell.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Rendered content */
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node) : _propTypes2.default.any).apply(this, arguments);
  },

  /** Rendered element */
  element: _propTypes2.default.string,

  /** See Table's Column type */
  label: _propTypes2.default.string,

  /** See Table's Column type */
  minWidth: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),

  /** See Table's Column type */
  maxWidth: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),

  /**
   * Provides custom rendering control. See the
   * [custom header cell example](/components/table/#custom-header-cell)
   * and our [render props guide](/render-props).
   */
  render: function render() {
    return (typeof _Table.bpfrpt_proptype_RenderFn === 'function' ? _Table.bpfrpt_proptype_RenderFn : _propTypes2.default.shape(_Table.bpfrpt_proptype_RenderFn)).apply(this, arguments);
  },

  /** See Table's Column type */
  textAlign: _propTypes2.default.oneOf(['start', 'end', 'center', 'justify']),

  /** See Table's Column type */
  width: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string])
} : {};
