// mineral-ui/Table/TableRow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utils = require('../utils');

var _styles = require('../styles');

var _TableBase = require('./TableBase');

var _Table = require('./Table');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// prettier-ignore
var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    TableRow_backgroundColor_highContrast_selected: baseTheme.color_theme_20,
    TableRow_backgroundColor_highContrast_selectedHover: baseTheme.color_theme_30,
    TableRow_backgroundColor_hover: baseTheme.color_gray_20,
    TableRow_backgroundColor_selected: baseTheme.color_theme_10,
    TableRow_backgroundColor_selectedHover: baseTheme.color_theme_20,
    TableRow_backgroundColor_striped: baseTheme.color_gray_10,
    TableRow_borderHorizontal: '1px solid ' + baseTheme.color_white,
    TableRow_borderHorizontal_highContrast: '1px solid ' + baseTheme.color_gray_60

  }, baseTheme);
};

var Root = (0, _styles.createStyledComponent)('tr', function (_ref) {
  var highContrast = _ref.highContrast,
      isSelected = _ref.isSelected,
      baseTheme = _ref.theme,
      striped = _ref.striped;

  var theme = componentTheme(baseTheme);

  return _extends({
    backgroundColor: function () {
      if (isSelected) {
        if (highContrast) {
          return theme.TableRow_backgroundColor_highContrast_selected;
        }
        return theme.TableRow_backgroundColor_selected;
      }
    }()

  }, highContrast ? {
    borderBottom: theme.TableRow_borderHorizontal_highContrast
  } : {
    '&:not(:last-child)': {
      borderBottom: theme.TableRow_borderHorizontal
    }
  }, {

    '*:not(thead) > &:hover': {
      backgroundColor: function () {
        if (isSelected) {
          if (highContrast) {
            return theme.TableRow_backgroundColor_highContrast_selectedHover;
          }
          return theme.TableRow_backgroundColor_selectedHover;
        }
        return theme.TableRow_backgroundColor_hover;
      }()
    },

    '&:nth-child(even):not(:hover)': {
      backgroundColor: !isSelected && striped ? theme.TableRow_backgroundColor_striped : null
    }

  }, isSelected ? {
    '& > td:first-child, & > th:first-child': {
      position: 'relative',

      '&::before': {
        backgroundColor: theme.color_theme_60,
        bottom: 0,
        content: '""',
        left: theme.direction !== 'rtl' ? 0 : null,
        right: theme.direction === 'rtl' ? 0 : null,
        position: 'absolute',
        top: 0,
        width: '4px'
      }
    }
  } : undefined);
}, {
  displayName: 'TableRow',
  rootEl: 'tr'
});

/**
 * TableRow
 */

var TableRow = function (_PureComponent) {
  _inherits(TableRow, _PureComponent);

  function TableRow() {
    _classCallCheck(this, TableRow);

    return _possibleConstructorReturn(this, (TableRow.__proto__ || Object.getPrototypeOf(TableRow)).apply(this, arguments));
  }

  _createClass(TableRow, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _jsx(_TableBase.TableContext.Consumer, {}, void 0, function (tableContextProps) {
        var _props = _this2.props,
            render = _props.render,
            restProps = _objectWithoutProperties(_props, ['render']);

        var rootProps = _extends({}, tableContextProps, restProps);

        if ((0, _utils.isRenderProp)(render)) {
          return render({
            props: rootProps
          });
        }

        return _react2.default.createElement(Root, rootProps);
      });
    }
  }]);

  return TableRow;
}(_react.PureComponent);

exports.default = TableRow;
TableRow.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Rendered content can be TD or TH */
  children: function children() {
    return (typeof React$Node === 'function' ? _propTypes2.default.instanceOf(React$Node).isRequired : _propTypes2.default.any.isRequired).apply(this, arguments);
  },

  /** Renders row with selected styles */
  isSelected: _propTypes2.default.bool,

  /** @Private Provided to user in render prop callback */
  isSelectable: _propTypes2.default.bool,

  /**
   * Provides custom rendering control. See the
   * [custom row example](/components/table/#custom-row) and
   * our [render props guide](/render-props).
   */
  render: function render() {
    return (typeof _Table.bpfrpt_proptype_RenderFn === 'function' ? _Table.bpfrpt_proptype_RenderFn : _propTypes2.default.shape(_Table.bpfrpt_proptype_RenderFn)).apply(this, arguments);
  },

  /** Row index */
  rowIndex: _propTypes2.default.number
} : {};
