// @arranger/admin/dist/schemas/ProjectSchema/resolvers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const utils_1 = require("./utils");
const projectsQueryResolver = (_, args, { es }, info) => tslib_1.__awaiter(void 0, void 0, void 0, function* () { return utils_1.getArrangerProjects(es); });
const singleProjectQueryResolver = (_, { id }, { es }, info) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const projects = yield utils_1.getArrangerProjects(es);
    return projects.find(({ id: _id }) => id === _id);
});
const newProjectMutationResolver = (_, { id }, { es }, info) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    return utils_1.addArrangerProject(es)(id).catch((err) => {
        err.message = 'potential project ID conflict';
        return Promise.reject(err);
    });
});
const deleteProjectMutationResolver = (_, { id }, { es }, info) => tslib_1.__awaiter(void 0, void 0, void 0, function* () { return utils_1.removeArrangerProject(es)(id); });
exports.createResolvers = () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    return ({
        Query: {
            projects: projectsQueryResolver,
            project: singleProjectQueryResolver,
        },
        Mutation: {
            newProject: newProjectMutationResolver,
            deleteProject: deleteProjectMutationResolver,
        },
    });
});
