// @arranger/middleware/dist/buildAggregations/injectNestedFiltersToAggs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _buildQuery = require("../buildQuery");

var _normalizeFilters = _interopRequireDefault(require("../buildQuery/normalizeFilters"));

var _constants = require("../constants");

var _lodash = require("lodash");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/*
 * due to this problem: https://github.com/kids-first/kf-portal-ui/issues/488
 * queries that are on a term that shares a parent with a aggregation field
 * needs to be dropped down to the aggregation level as a filter.
 */
var injectNestedFiltersToAggs = function injectNestedFiltersToAggs(_ref) {
  var aggs = _ref.aggs,
      nestedSqonFilters = _ref.nestedSqonFilters,
      aggregationsFilterThemselves = _ref.aggregationsFilterThemselves;
  return Object.entries(aggs).reduce(function (acc, _ref2) {
    var _ref3 = _slicedToArray(_ref2, 2),
        aggName = _ref3[0],
        aggContent = _ref3[1];

    var skipToNextLevel = function skipToNextLevel() {
      acc[aggName] = _objectSpread({}, aggContent, {
        aggs: injectNestedFiltersToAggs({
          aggs: aggContent.aggs,
          nestedSqonFilters: nestedSqonFilters,
          aggregationsFilterThemselves: aggregationsFilterThemselves
        })
      });
      return acc;
    };

    var wrapInFilterAgg = function wrapInFilterAgg() {
      acc[aggName] = _objectSpread({}, aggContent, {
        aggs: _defineProperty({}, "".concat(aggContent.nested.path, ":").concat(_constants.AGGS_WRAPPER_FILTERED), {
          filter: {
            bool: {
              should: nestedSqonFilters[aggContent.nested.path].filter(function (sqonFilter) {
                return aggregationsFilterThemselves || aggName.split(':')[0] !== sqonFilter.content.field;
              }).map(function (sqonFilter) {
                return (0, _buildQuery.opSwitch)({
                  nestedFields: [],
                  filter: (0, _normalizeFilters["default"])(sqonFilter)
                });
              })
            }
          },
          aggs: injectNestedFiltersToAggs({
            aggs: aggContent.aggs,
            nestedSqonFilters: nestedSqonFilters,
            aggregationsFilterThemselves: aggregationsFilterThemselves
          })
        })
      });
      return acc;
    };

    if (aggContent.global || aggContent.filter) {
      return skipToNextLevel();
    } else if (aggContent.nested) {
      if (nestedSqonFilters[aggContent.nested.path]) {
        return wrapInFilterAgg();
      } else {
        return skipToNextLevel();
      }
    } else {
      return acc;
    }
  }, (0, _lodash.cloneDeep)(aggs));
};

var _default = injectNestedFiltersToAggs;
exports["default"] = _default;
