// mineral-ui/es/Portal/Portal.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import { Children, Component } from 'react';
import { canUseDOM } from 'exenv';
import { unmountComponentAtNode, unstable_renderSubtreeIntoContainer } from 'react-dom';

var Portal = function (_Component) {
  _inherits(Portal, _Component);

  function Portal() {
    _classCallCheck(this, Portal);

    return _possibleConstructorReturn(this, (Portal.__proto__ || Object.getPrototypeOf(Portal)).apply(this, arguments));
  }

  _createClass(Portal, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.createPortal();
      this.renderIntoPortal();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.renderIntoPortal();
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.closePortal();
    }
  }, {
    key: 'render',
    value: function render() {
      return null;
    }
  }, {
    key: 'renderIntoPortal',
    value: function renderIntoPortal() {
      if (canUseDOM) {
        var _props = this.props,
            _children = _props.children,
            _callback = _props.callback;

        var content = Children.only(_children);

        this.content = unstable_renderSubtreeIntoContainer(this, content, this.node, _callback);
      }
    }
  }, {
    key: 'createPortal',
    value: function createPortal() {
      if (canUseDOM) {
        this.node = global.document.createElement('div');
        global.document.body.appendChild(this.node);
      }
    }
  }, {
    key: 'closePortal',
    value: function closePortal() {
      if (canUseDOM) {
        unmountComponentAtNode(this.node);
        if (global.document.body.contains && global.document.body.contains(this.node)) {
          global.document.body.removeChild(this.node);
        }
      }
    }
  }]);

  return Portal;
}(Component);

export default Portal;
Portal.propTypes = process.env.NODE_ENV !== "production" ? {
  children: function children() {
    return (typeof React$Node === 'function' ? PropTypes.instanceOf(React$Node).isRequired : PropTypes.any.isRequired).apply(this, arguments);
  },
  callback: PropTypes.func
} : {};
import PropTypes from 'prop-types';
