// @arranger/middleware/dist/buildAggregations/createFieldAggregation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _lodash = require("lodash");

var _constants = require("../constants");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var MAX_AGGREGATION_SIZE = 300000;
var HISTOGRAM_INTERVAL_DEFAULT = 1000;

var createNumericAggregation = function createNumericAggregation(_ref) {
  var type = _ref.type,
      field = _ref.field,
      graphqlField = _ref.graphqlField;
  var args = (0, _lodash.get)(graphqlField, [type, '__arguments', 0]) || {};
  return _defineProperty({}, "".concat(field, ":").concat(type), _defineProperty({}, type, _objectSpread({
    field: field
  }, type === _constants.HISTOGRAM ? {
    interval: (0, _lodash.get)(args, 'interval.value') || HISTOGRAM_INTERVAL_DEFAULT
  } : {})));
};

var createTermAggregation = function createTermAggregation(_ref3) {
  var _ref4;

  var field = _ref3.field,
      isNested = _ref3.isNested;
  return _ref4 = {}, _defineProperty(_ref4, field, _objectSpread({}, isNested ? {
    aggs: {
      rn: {
        reverse_nested: {}
      }
    }
  } : {}, {
    terms: {
      field: field,
      size: MAX_AGGREGATION_SIZE
    }
  })), _defineProperty(_ref4, "".concat(field, ":missing"), _objectSpread({}, isNested ? {
    aggs: {
      rn: {
        reverse_nested: {}
      }
    }
  } : {}, {
    missing: {
      field: field
    }
  })), _ref4;
};
/**
 * graphqlFields: output from `graphql-fields` (https://github.com/robrichard/graphql-fields)
 */


var _default = function _default(_ref5) {
  var field = _ref5.field,
      _ref5$graphqlField = _ref5.graphqlField,
      graphqlField = _ref5$graphqlField === void 0 ? {} : _ref5$graphqlField,
      _ref5$isNested = _ref5.isNested,
      isNested = _ref5$isNested === void 0 ? false : _ref5$isNested;
  var types = [_constants.BUCKETS, _constants.STATS, _constants.HISTOGRAM].filter(function (t) {
    return graphqlField[t];
  });
  return types.reduce(function (acc, type) {
    if (type === _constants.BUCKETS) {
      return Object.assign(acc, createTermAggregation({
        field: field,
        isNested: isNested
      }));
    } else if ([_constants.STATS, _constants.HISTOGRAM].includes(type)) {
      return Object.assign(acc, createNumericAggregation({
        type: type,
        field: field,
        graphqlField: graphqlField
      }));
    } else {
      return acc;
    }
  }, {});
};

exports["default"] = _default;
