// @arranger/middleware/dist/buildAggregations/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _default;

var _lodash = require("lodash");

var _injectNestedFiltersToAggs = _interopRequireDefault(require("./injectNestedFiltersToAggs"));

var _getNestedSqonFilters = _interopRequireDefault(require("./getNestedSqonFilters"));

var _constants = require("../constants");

var _createFieldAggregation = _interopRequireDefault(require("./createFieldAggregation"));

var _normalizeFilters = _interopRequireDefault(require("../buildQuery/normalizeFilters"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function createGlobalAggregation(_ref) {
  var field = _ref.field,
      aggregation = _ref.aggregation;
  return _defineProperty({}, "".concat(field, ":").concat(_constants.AGGS_WRAPPER_GLOBAL), {
    global: {},
    aggs: aggregation
  });
}

function createFilteredAggregation(_ref3) {
  var field = _ref3.field,
      filter = _ref3.filter,
      aggregation = _ref3.aggregation;
  return Object.keys(filter || {}).length ? _defineProperty({}, "".concat(field, ":").concat(_constants.AGGS_WRAPPER_FILTERED), {
    filter: filter,
    aggs: aggregation
  }) : aggregation;
}

function removeFieldFromQuery(_ref5) {
  var field = _ref5.field,
      query = _ref5.query;
  var nested = (0, _lodash.get)(query, _constants.ES_NESTED);
  var nestedQuery = (0, _lodash.get)(nested, _constants.ES_QUERY);
  var bool = (0, _lodash.get)(query, _constants.ES_BOOL);

  if (['terms', 'range'].some(function (k) {
    return (0, _lodash.get)(query, [k, field]);
  }) || (0, _lodash.get)(query, ['exists', 'field']) === field) {
    return null;
  } else if (nestedQuery) {
    var cleaned = removeFieldFromQuery({
      field: field,
      query: nestedQuery
    });
    return cleaned && _objectSpread({}, query, _defineProperty({}, _constants.ES_NESTED, _objectSpread({}, nested, _defineProperty({}, _constants.ES_QUERY, cleaned))));
  } else if (bool) {
    var filtered = Object.entries(bool).reduce(function (acc, _ref6) {
      var _ref7 = _slicedToArray(_ref6, 2),
          type = _ref7[0],
          values = _ref7[1];

      var filteredValues = values.map(function (value) {
        return removeFieldFromQuery({
          field: field,
          query: value
        });
      }).filter(Boolean);

      if (filteredValues.length > 0) {
        acc[type] = filteredValues;
      }

      return acc;
    }, {});
    return Object.keys(filtered).length > 0 ? _defineProperty({}, _constants.ES_BOOL, filtered) : null;
  } else {
    return query;
  }
}

function getNestedPathsInField(_ref9) {
  var field = _ref9.field,
      nestedFields = _ref9.nestedFields;
  return field.split('.').map(function (s, i, arr) {
    return arr.slice(0, i + 1).join('.');
  }).filter(function (p) {
    return nestedFields.includes(p);
  });
}

function wrapWithFilters(_ref10) {
  var field = _ref10.field,
      query = _ref10.query,
      aggregationsFilterThemselves = _ref10.aggregationsFilterThemselves,
      aggregation = _ref10.aggregation;

  if (!aggregationsFilterThemselves) {
    var cleanedQuery = removeFieldFromQuery({
      field: field,
      query: query
    }); // TODO: better way to figure out that the field wasn't found

    if (!(0, _lodash.isEqual)(cleanedQuery || {}, query || {})) {
      return createGlobalAggregation({
        field: field,
        aggregation: createFilteredAggregation({
          field: field,
          filter: cleanedQuery,
          aggregation: aggregation
        })
      });
    }
  }

  return aggregation;
}
/**
 * graphqlFields: output from `graphql-fields` (https://github.com/robrichard/graphql-fields)
 */


function _default(_ref11) {
  var sqon = _ref11.sqon,
      graphqlFields = _ref11.graphqlFields,
      nestedFields = _ref11.nestedFields,
      aggregationsFilterThemselves = _ref11.aggregationsFilterThemselves,
      query = _ref11.query;
  var normalizedSqon = (0, _normalizeFilters["default"])(sqon);
  var nestedSqonFilters = (0, _getNestedSqonFilters["default"])({
    sqon: normalizedSqon,
    nestedFields: nestedFields
  });
  var aggs = Object.entries(graphqlFields).reduce(function (aggregations, _ref12) {
    var _ref13 = _slicedToArray(_ref12, 2),
        fieldKey = _ref13[0],
        graphqlField = _ref13[1];

    var field = fieldKey.replace(/__/g, '.');
    var nestedPaths = getNestedPathsInField({
      field: field,
      nestedFields: nestedFields
    });
    var fieldAggregation = (0, _createFieldAggregation["default"])({
      field: field,
      graphqlField: graphqlField,
      isNested: nestedPaths.length
    });
    var aggregation = nestedPaths.reverse().reduce(function (aggs, path) {
      return _defineProperty({}, "".concat(field, ":").concat(_constants.AGGS_WRAPPER_NESTED), {
        nested: {
          path: path
        },
        aggs: aggs
      });
    }, fieldAggregation);
    return Object.assign(aggregations, wrapWithFilters({
      query: query,
      field: field,
      aggregation: aggregation,
      aggregationsFilterThemselves: aggregationsFilterThemselves
    }));
  }, {});
  var filteredAggregations = (0, _injectNestedFiltersToAggs["default"])({
    aggs: aggs,
    nestedSqonFilters: nestedSqonFilters,
    aggregationsFilterThemselves: aggregationsFilterThemselves
  });
  return filteredAggregations;
}
