// mineral-ui/es/utils/isValidProp.js
import _htmlAttributes from 'react-html-attributes';
import memoize from 'fast-memoize';
import reactProps from './reactProps';
import { settify } from './collections';

var REGEX_DATA_OR_ARIA = /^(data|aria)-/;
var htmlAttributes = settify(_htmlAttributes);
var globalHtmlAttributes = htmlAttributes['*'] || new Set();
var svgTags = htmlAttributes.elements ? htmlAttributes.elements.svg : new Set();
var svgAttributes = htmlAttributes.svg || new Set();

var isElement = function isElement(tag) {
  return typeof tag === 'string';
};
var isComponent = function isComponent(tag) {
  return !isElement(tag);
};
var isReactProp = function isReactProp(prop) {
  return reactProps[prop];
};
var isCustomAttribute = function isCustomAttribute(prop) {
  return REGEX_DATA_OR_ARIA.test(prop);
};
var isSvgAttribute = function isSvgAttribute(tag, prop) {
  return svgTags.has(tag) && svgAttributes.has(prop);
};
var isHtmlAttribute = function isHtmlAttribute(tag, prop) {
  return globalHtmlAttributes.has(prop) || htmlAttributes[tag] && htmlAttributes[tag].has(prop);
};

var isValidProp = function isValidProp(tag, prop) {
  return isComponent(tag) || isReactProp(prop) || isHtmlAttribute(tag, prop) || isSvgAttribute(tag, prop) || isCustomAttribute(prop);
};

export default memoize(isValidProp);
