// @arranger/components/dist/Aggs/DatesAgg.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _emotion = require("emotion");

var _react = _interopRequireDefault(require("react"));

var _moment = _interopRequireDefault(require("moment"));

var _reactDatepicker = _interopRequireDefault(require("react-datepicker"));

require("react-datepicker/dist/react-datepicker.css");

var _utils = require("../SQONView/utils");

require("./AggregationCard.css");

var _AggsWrapper = _interopRequireDefault(require("./AggsWrapper"));

require("./DatesAgg.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var SQON_DATE_FORMAT = 'YYYY-MM-DD';

var dateFromSqon = function dateFromSqon(dateString) {
  return (0, _moment.default)(dateString, SQON_DATE_FORMAT);
};

var momentToSqonDate = function momentToSqonDate(moment) {
  return moment === null || moment === void 0 ? void 0 : moment.format(SQON_DATE_FORMAT);
};

var DatesAgg =
/*#__PURE__*/
function (_React$Component) {
  _inherits(DatesAgg, _React$Component);

  function DatesAgg(props) {
    var _this;

    _classCallCheck(this, DatesAgg);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(DatesAgg).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "initializeState", function (_ref) {
      var _ref$stats = _ref.stats,
          stats = _ref$stats === void 0 ? {} : _ref$stats,
          _ref$getActiveValue = _ref.getActiveValue,
          getActiveValue = _ref$getActiveValue === void 0 ? function () {
        return null;
      } : _ref$getActiveValue;
      var field = _this.props.field;
      var minDate = stats.min && (0, _moment.default)(stats.min).subtract(1, 'days');
      var maxDate = stats.max && (0, _moment.default)(stats.max).add(1, 'days');
      var startFromSqon = getActiveValue({
        op: '>=',
        field: field
      });
      var endFromSqon = getActiveValue({
        op: '<=',
        field: field
      });
      return {
        minDate: minDate,
        maxDate: maxDate,
        startDate: startFromSqon ? dateFromSqon(startFromSqon) : null,
        endDate: endFromSqon ? dateFromSqon(endFromSqon) : null
      };
    });

    _defineProperty(_assertThisInitialized(_this), "updateSqon", function () {
      var _this$state = _this.state,
          startDate = _this$state.startDate,
          endDate = _this$state.endDate;
      var _this$props = _this.props,
          field = _this$props.field,
          handleDateChange = _this$props.handleDateChange;

      if (handleDateChange && field) {
        var content = [].concat(_toConsumableArray(startDate ? [{
          op: '>=',
          content: {
            field: field,
            value: momentToSqonDate(startDate.startOf('day'))
          }
        }] : []), _toConsumableArray(endDate ? [{
          op: '<=',
          content: {
            field: field,
            value: momentToSqonDate(endDate.endOf('day'))
          }
        }] : []));
        handleDateChange({
          field: field,
          value: content,
          generateNextSQON: function generateNextSQON(sqon) {
            return (0, _utils.replaceSQON)(content.length ? {
              op: 'and',
              content: content
            } : null, (0, _utils.removeSQON)(field, sqon));
          }
        });
      }
    });

    _this.state = _this.initializeState(props);
    return _this;
  }

  _createClass(DatesAgg, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      this.setState(this.initializeState(nextProps));
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          _this$props2$displayN = _this$props2.displayName,
          displayName = _this$props2$displayN === void 0 ? 'Date Range' : _this$props2$displayN,
          _this$props2$collapsi = _this$props2.collapsible,
          collapsible = _this$props2$collapsi === void 0 ? true : _this$props2$collapsi,
          WrapperComponent = _this$props2.WrapperComponent,
          _this$props2$facetVie = _this$props2.facetView,
          facetView = _this$props2$facetVie === void 0 ? false : _this$props2$facetVie;
      var _this$state2 = this.state,
          minDate = _this$state2.minDate,
          maxDate = _this$state2.maxDate,
          startDate = _this$state2.startDate,
          endDate = _this$state2.endDate;
      return _react.default.createElement(_AggsWrapper.default, {
        displayName: displayName,
        WrapperComponent: WrapperComponent,
        collapsible: collapsible
      }, _react.default.createElement("div", {
        className:
        /*#__PURE__*/
        (0, _emotion.css)("display:flex;align-items:center;justify-content:space-around;")
      }, _react.default.createElement(_reactDatepicker.default, _extends({
        minDate: minDate,
        maxDate: maxDate
      }, {
        isClearable: true,
        openToDate: minDate,
        popperPlacement: facetView ? 'bottom-start' : 'top-start',
        disabledKeyboardNavigation: true,
        placeholderText: "Start Date",
        selected: startDate,
        onChange: function onChange(x) {
          return _this2.setState({
            startDate: x
          }, _this2.updateSqon);
        },
        "aria-label": "Pick start date"
      })), _react.default.createElement(_reactDatepicker.default, _extends({
        minDate: minDate,
        maxDate: maxDate
      }, {
        isClearable: true,
        openToDate: maxDate,
        popperPlacement: facetView ? 'bottom-end' : 'top-start',
        disabledKeyboardNavigation: true,
        placeholderText: "End Date",
        selected: endDate,
        onChange: function onChange(x) {
          return _this2.setState({
            endDate: x
          }, _this2.updateSqon);
        },
        "aria-label": "Pick end date"
      }))));
    }
  }]);

  return DatesAgg;
}(_react.default.Component);

var _default = DatesAgg;
exports.default = _default;
