// mineral-ui-tokens/all.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var backgroundColor = exports.backgroundColor = '#ffffff';
var backgroundColor_active = exports.backgroundColor_active = '#ebeff5';
var backgroundColor_disabled = exports.backgroundColor_disabled = '#ebeff5';
var backgroundColor_focus = exports.backgroundColor_focus = '#ffffff';
var backgroundColor_hover = exports.backgroundColor_hover = '#f5f7fa';
var backgroundColor_brand_selected = exports.backgroundColor_brand_selected = '#f0f5fc';
var backgroundColor_brand_selectedActive = exports.backgroundColor_brand_selectedActive = '#accbfc';
var backgroundColor_brand_selectedHover = exports.backgroundColor_brand_selectedHover = '#cfe0fc';
var backgroundColor_brandPrimary = exports.backgroundColor_brandPrimary = '#3272d9';
var backgroundColor_brandPrimary_active = exports.backgroundColor_brandPrimary_active = '#1d5bbf';
var backgroundColor_brandPrimary_focus = exports.backgroundColor_brandPrimary_focus = '#3272d9';
var backgroundColor_brandPrimary_hover = exports.backgroundColor_brandPrimary_hover = '#5691f0';
var backgroundColor_danger_active = exports.backgroundColor_danger_active = '#fad4d4';
var backgroundColor_danger_focus = exports.backgroundColor_danger_focus = '#faf0f0';
var backgroundColor_danger_hover = exports.backgroundColor_danger_hover = '#faf0f0';
var backgroundColor_dangerPrimary = exports.backgroundColor_dangerPrimary = '#de1b1b';
var backgroundColor_dangerPrimary_active = exports.backgroundColor_dangerPrimary_active = '#b80d0d';
var backgroundColor_dangerPrimary_focus = exports.backgroundColor_dangerPrimary_focus = '#de1b1b';
var backgroundColor_dangerPrimary_hover = exports.backgroundColor_dangerPrimary_hover = '#f55353';
var backgroundColor_success_active = exports.backgroundColor_success_active = '#abedc5';
var backgroundColor_success_focus = exports.backgroundColor_success_focus = '#e1faeb';
var backgroundColor_success_hover = exports.backgroundColor_success_hover = '#e1faeb';
var backgroundColor_successPrimary = exports.backgroundColor_successPrimary = '#2a854e';
var backgroundColor_successPrimary_active = exports.backgroundColor_successPrimary_active = '#20693d';
var backgroundColor_successPrimary_focus = exports.backgroundColor_successPrimary_focus = '#2a854e';
var backgroundColor_successPrimary_hover = exports.backgroundColor_successPrimary_hover = '#3ba164';
var backgroundColor_warning_active = exports.backgroundColor_warning_active = '#fad8af';
var backgroundColor_warning_focus = exports.backgroundColor_warning_focus = '#fcf2e6';
var backgroundColor_warning_hover = exports.backgroundColor_warning_hover = '#fcf2e6';
var backgroundColor_warningPrimary = exports.backgroundColor_warningPrimary = '#ad5f00';
var backgroundColor_warningPrimary_active = exports.backgroundColor_warningPrimary_active = '#8a4d03';
var backgroundColor_warningPrimary_focus = exports.backgroundColor_warningPrimary_focus = '#ad5f00';
var backgroundColor_warningPrimary_hover = exports.backgroundColor_warningPrimary_hover = '#cf7911';
var borderColor = exports.borderColor = '#c8d1e0';
var borderColor_brand = exports.borderColor_brand = '#3272d9';
var borderColor_brand_active = exports.borderColor_brand_active = '#1d5bbf';
var borderColor_brand_focus = exports.borderColor_brand_focus = '#1d5bbf';
var borderColor_brand_hover = exports.borderColor_brand_hover = '#5691f0';
var borderColor_danger = exports.borderColor_danger = '#de1b1b';
var borderColor_danger_active = exports.borderColor_danger_active = '#b80d0d';
var borderColor_danger_focus = exports.borderColor_danger_focus = '#b80d0d';
var borderColor_danger_hover = exports.borderColor_danger_hover = '#f55353';
var borderColor_success = exports.borderColor_success = '#2a854e';
var borderColor_success_active = exports.borderColor_success_active = '#20693d';
var borderColor_success_focus = exports.borderColor_success_focus = '#20693d';
var borderColor_success_hover = exports.borderColor_success_hover = '#3ba164';
var borderColor_warning = exports.borderColor_warning = '#ad5f00';
var borderColor_warning_active = exports.borderColor_warning_active = '#8a4d03';
var borderColor_warning_focus = exports.borderColor_warning_focus = '#8a4d03';
var borderColor_warning_hover = exports.borderColor_warning_hover = '#cf7911';
var borderRadius_1 = exports.borderRadius_1 = '3px';
var boxShadow_1 = exports.boxShadow_1 = '0 1px 2px 0 rgba(0,0,0,0.2), 0 2px 4px 0 rgba(0,0,0,0.2)';
var boxShadow_2 = exports.boxShadow_2 = '0 2px 4px 0 rgba(0,0,0,0.2), 0 4px 8px 0 rgba(0,0,0,0.2)';
var boxShadow_3 = exports.boxShadow_3 = '0 4px 8px 0 rgba(0,0,0,0.2), 0 8px 16px 0 rgba(0,0,0,0.2)';
var boxShadow_4 = exports.boxShadow_4 = '0 8px 16px 0 rgba(0,0,0,0.2), 0 20px 16px -8px rgba(0,0,0,0.2)';
var boxShadow_5 = exports.boxShadow_5 = '0 16px 24px 0 rgba(0,0,0,0.2), 0 32px 24px -16px rgba(0,0,0,0.2)';
var breakpoint_narrow = exports.breakpoint_narrow = '512px';
var breakpoint_medium = exports.breakpoint_medium = '768px';
var breakpoint_wide = exports.breakpoint_wide = '1024px';
var color = exports.color = '#333840';
var color_inverted = exports.color_inverted = '#ffffff';
var color_disabled = exports.color_disabled = '#afbacc';
var color_mouse = exports.color_mouse = '#58606e';
var color_readOnly = exports.color_readOnly = '#58606e';
var color_required = exports.color_required = '#de1b1b';
var color_brand = exports.color_brand = '#3272d9';
var color_brandPrimary = exports.color_brandPrimary = '#ffffff';
var color_brand_active = exports.color_brand_active = '#1d5bbf';
var color_brand_focus = exports.color_brand_focus = '#3272d9';
var color_brand_hover = exports.color_brand_hover = '#5691f0';
var color_danger = exports.color_danger = '#de1b1b';
var color_dangerPrimary = exports.color_dangerPrimary = '#ffffff';
var color_danger_active = exports.color_danger_active = '#b80d0d';
var color_danger_focus = exports.color_danger_focus = '#de1b1b';
var color_danger_hover = exports.color_danger_hover = '#f55353';
var color_success = exports.color_success = '#2a854e';
var color_successPrimary = exports.color_successPrimary = '#ffffff';
var color_success_active = exports.color_success_active = '#20693d';
var color_success_focus = exports.color_success_focus = '#2a854e';
var color_success_hover = exports.color_success_hover = '#3ba164';
var color_warning = exports.color_warning = '#ad5f00';
var color_warningPrimary = exports.color_warningPrimary = '#ffffff';
var color_warning_active = exports.color_warning_active = '#8a4d03';
var color_warning_focus = exports.color_warning_focus = '#ad5f00';
var color_warning_hover = exports.color_warning_hover = '#cf7911';
var fontFamily = exports.fontFamily = '"Open Sans"';
var fontFamily_system = exports.fontFamily_system = '-apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"';
var fontFamily_monospace = exports.fontFamily_monospace = '"SF Mono", "Droid Sans Mono", "Source Code Pro", Monaco, Consolas, "Courier New", Courier, monospace';
var fontSize_base = exports.fontSize_base = '1rem';
var fontSize_mouse = exports.fontSize_mouse = '0.6875rem';
var fontSize_prose = exports.fontSize_prose = '1rem';
var fontSize_ui = exports.fontSize_ui = '0.875rem';
var fontWeight_regular = exports.fontWeight_regular = 400;
var fontWeight_semiBold = exports.fontWeight_semiBold = 600;
var fontWeight_bold = exports.fontWeight_bold = 700;
var fontWeight_extraBold = exports.fontWeight_extraBold = 800;
var lineHeight = exports.lineHeight = 1.25;
var lineHeight_heading = exports.lineHeight_heading = 1.25;
var lineHeight_heading_small = exports.lineHeight_heading_small = 1.5;
var lineHeight_prose = exports.lineHeight_prose = 1.5;
var size_small = exports.size_small = '24px';
var size_medium = exports.size_medium = '32px';
var size_large = exports.size_large = '40px';
var size_jumbo = exports.size_jumbo = '52px';
var space_inline_xxs = exports.space_inline_xxs = '2px';
var space_inline_xs = exports.space_inline_xs = '4px';
var space_inline_sm = exports.space_inline_sm = '8px';
var space_inline_md = exports.space_inline_md = '16px';
var space_inline_lg = exports.space_inline_lg = '24px';
var space_inline_xl = exports.space_inline_xl = '32px';
var space_inline_xxl = exports.space_inline_xxl = '64px';
var space_inset_sm = exports.space_inset_sm = '8px';
var space_inset_md = exports.space_inset_md = '16px';
var space_inset_lg = exports.space_inset_lg = '24px';
var space_stack_xxs = exports.space_stack_xxs = '2px';
var space_stack_xs = exports.space_stack_xs = '4px';
var space_stack_sm = exports.space_stack_sm = '8px';
var space_stack_md = exports.space_stack_md = '16px';
var space_stack_lg = exports.space_stack_lg = '24px';
var space_stack_xl = exports.space_stack_xl = '32px';
var space_stack_xxl = exports.space_stack_xxl = '64px';
var zIndex_100 = exports.zIndex_100 = 100;
var zIndex_200 = exports.zIndex_200 = 200;
var zIndex_400 = exports.zIndex_400 = 400;
var zIndex_800 = exports.zIndex_800 = 800;
var zIndex_1600 = exports.zIndex_1600 = 1600;
var h1_color = exports.h1_color = '#333840';
var h1_fontSize = exports.h1_fontSize = '2.125rem';
var h1_fontWeight = exports.h1_fontWeight = 800;
var h2_color = exports.h2_color = '#58606e';
var h2_fontSize = exports.h2_fontSize = '1.75rem';
var h2_fontWeight = exports.h2_fontWeight = 700;
var h3_color = exports.h3_color = '#58606e';
var h3_fontSize = exports.h3_fontSize = '1.375rem';
var h3_fontWeight = exports.h3_fontWeight = 700;
var h4_color = exports.h4_color = '#58606e';
var h4_fontSize = exports.h4_fontSize = '1.125rem';
var h4_fontWeight = exports.h4_fontWeight = 700;
var h5_color = exports.h5_color = '#333840';
var h5_fontSize = exports.h5_fontSize = '0.875rem';
var h5_fontWeight = exports.h5_fontWeight = 700;
var h6_color = exports.h6_color = '#58606e';
var h6_fontSize = exports.h6_fontSize = '0.875rem';
var h6_fontWeight = exports.h6_fontWeight = 400;
var icon_color = exports.icon_color = '#58606e';
var icon_color_brand = exports.icon_color_brand = '#3272d9';
var icon_color_danger = exports.icon_color_danger = '#de1b1b';
var icon_color_success = exports.icon_color_success = '#2a854e';
var icon_color_warning = exports.icon_color_warning = '#ad5f00';
var input_backgroundColor = exports.input_backgroundColor = '#ffffff';
var input_backgroundColor_disabled = exports.input_backgroundColor_disabled = '#ebeff5';
var input_color_placeholder = exports.input_color_placeholder = '#8e99ab';
var panel_backgroundColor = exports.panel_backgroundColor = '#ffffff';
var panel_backgroundColor_inverted = exports.panel_backgroundColor_inverted = '#434a54';
var panel_borderColor = exports.panel_borderColor = '#ebeff5';
var panel_borderColor_inverted = exports.panel_borderColor_inverted = '#434a54';
var well_backgroundColor = exports.well_backgroundColor = '#ebeff5';
var well_backgroundColor_danger = exports.well_backgroundColor_danger = '#fad4d4';
var well_backgroundColor_success = exports.well_backgroundColor_success = '#abedc5';
var well_backgroundColor_warning = exports.well_backgroundColor_warning = '#fad8af';
var well_borderColor_danger = exports.well_borderColor_danger = '#fa8e8e';
var well_borderColor_success = exports.well_borderColor_success = '#57c282';
var well_borderColor_warning = exports.well_borderColor_warning = '#e89c3f';
var black = exports.black = '#1d1f24';
var white = exports.white = '#ffffff';
var brand_10 = exports.brand_10 = '#f0f5fc';
var brand_20 = exports.brand_20 = '#cfe0fc';
var brand_30 = exports.brand_30 = '#accbfc';
var brand_40 = exports.brand_40 = '#84b1fa';
var brand_50 = exports.brand_50 = '#5691f0';
var brand_60 = exports.brand_60 = '#3272d9';
var brand_70 = exports.brand_70 = '#1d5bbf';
var brand_80 = exports.brand_80 = '#114599';
var brand_90 = exports.brand_90 = '#103570';
var brand_100 = exports.brand_100 = '#15233b';
var blue_10 = exports.blue_10 = '#f0f5fc';
var blue_20 = exports.blue_20 = '#cfe0fc';
var blue_30 = exports.blue_30 = '#accbfc';
var blue_40 = exports.blue_40 = '#84b1fa';
var blue_50 = exports.blue_50 = '#5691f0';
var blue_60 = exports.blue_60 = '#3272d9';
var blue_70 = exports.blue_70 = '#1d5bbf';
var blue_80 = exports.blue_80 = '#114599';
var blue_90 = exports.blue_90 = '#103570';
var blue_100 = exports.blue_100 = '#15233b';
var bronze_10 = exports.bronze_10 = '#fcf2e6';
var bronze_20 = exports.bronze_20 = '#fad8af';
var bronze_30 = exports.bronze_30 = '#f5bc76';
var bronze_40 = exports.bronze_40 = '#e89c3f';
var bronze_50 = exports.bronze_50 = '#cf7911';
var bronze_60 = exports.bronze_60 = '#ad5f00';
var bronze_70 = exports.bronze_70 = '#8a4d03';
var bronze_80 = exports.bronze_80 = '#693d07';
var bronze_90 = exports.bronze_90 = '#4d2f0b';
var bronze_100 = exports.bronze_100 = '#33210c';
var dusk_10 = exports.dusk_10 = '#f4f2f7';
var dusk_20 = exports.dusk_20 = '#e3dcf7';
var dusk_30 = exports.dusk_30 = '#cec2f0';
var dusk_40 = exports.dusk_40 = '#b5a6e3';
var dusk_50 = exports.dusk_50 = '#9886cf';
var dusk_60 = exports.dusk_60 = '#7a68b3';
var dusk_70 = exports.dusk_70 = '#645396';
var dusk_80 = exports.dusk_80 = '#4f4178';
var dusk_90 = exports.dusk_90 = '#3a3154';
var dusk_100 = exports.dusk_100 = '#282436';
var gray_10 = exports.gray_10 = '#f5f7fa';
var gray_20 = exports.gray_20 = '#ebeff5';
var gray_30 = exports.gray_30 = '#dde3ed';
var gray_40 = exports.gray_40 = '#c8d1e0';
var gray_50 = exports.gray_50 = '#afbacc';
var gray_60 = exports.gray_60 = '#8e99ab';
var gray_70 = exports.gray_70 = '#707a8a';
var gray_80 = exports.gray_80 = '#58606e';
var gray_90 = exports.gray_90 = '#434a54';
var gray_100 = exports.gray_100 = '#333840';
var green_10 = exports.green_10 = '#e1faeb';
var green_20 = exports.green_20 = '#abedc5';
var green_30 = exports.green_30 = '#7ddba3';
var green_40 = exports.green_40 = '#57c282';
var green_50 = exports.green_50 = '#3ba164';
var green_60 = exports.green_60 = '#2a854e';
var green_70 = exports.green_70 = '#20693d';
var green_80 = exports.green_80 = '#1a5230';
var green_90 = exports.green_90 = '#153d25';
var green_100 = exports.green_100 = '#112b1b';
var indigo_10 = exports.indigo_10 = '#f2f2fc';
var indigo_20 = exports.indigo_20 = '#dcdcfc';
var indigo_30 = exports.indigo_30 = '#c2c2fc';
var indigo_40 = exports.indigo_40 = '#a7a7fa';
var indigo_50 = exports.indigo_50 = '#8585f2';
var indigo_60 = exports.indigo_60 = '#6767e6';
var indigo_70 = exports.indigo_70 = '#4d4dd1';
var indigo_80 = exports.indigo_80 = '#3737b3';
var indigo_90 = exports.indigo_90 = '#28288a';
var indigo_100 = exports.indigo_100 = '#202057';
var magenta_10 = exports.magenta_10 = '#faf0f4';
var magenta_20 = exports.magenta_20 = '#fad4e4';
var magenta_30 = exports.magenta_30 = '#fab4d1';
var magenta_40 = exports.magenta_40 = '#f78bb8';
var magenta_50 = exports.magenta_50 = '#ed5393';
var magenta_60 = exports.magenta_60 = '#d6246e';
var magenta_70 = exports.magenta_70 = '#b01355';
var magenta_80 = exports.magenta_80 = '#8a1244';
var magenta_90 = exports.magenta_90 = '#611535';
var magenta_100 = exports.magenta_100 = '#421527';
var purple_10 = exports.purple_10 = '#f5f0fa';
var purple_20 = exports.purple_20 = '#ead9fa';
var purple_30 = exports.purple_30 = '#dabcf7';
var purple_40 = exports.purple_40 = '#c79bf2';
var purple_50 = exports.purple_50 = '#ae74e8';
var purple_60 = exports.purple_60 = '#9656d6';
var purple_70 = exports.purple_70 = '#7d3cbd';
var purple_80 = exports.purple_80 = '#642b9e';
var purple_90 = exports.purple_90 = '#4b2175';
var purple_100 = exports.purple_100 = '#371c52';
var red_10 = exports.red_10 = '#faf0f0';
var red_20 = exports.red_20 = '#fad4d4';
var red_30 = exports.red_30 = '#fab6b6';
var red_40 = exports.red_40 = '#fa8e8e';
var red_50 = exports.red_50 = '#f55353';
var red_60 = exports.red_60 = '#de1b1b';
var red_70 = exports.red_70 = '#b80d0d';
var red_80 = exports.red_80 = '#8f0e0e';
var red_90 = exports.red_90 = '#661414';
var red_100 = exports.red_100 = '#451717';
var sky_10 = exports.sky_10 = '#e8f4fa';
var sky_20 = exports.sky_20 = '#bbe5fa';
var sky_30 = exports.sky_30 = '#8dd4f7';
var sky_40 = exports.sky_40 = '#53baed';
var sky_50 = exports.sky_50 = '#229ad6';
var sky_60 = exports.sky_60 = '#0c7bb3';
var sky_70 = exports.sky_70 = '#066391';
var sky_80 = exports.sky_80 = '#064d70';
var sky_90 = exports.sky_90 = '#093952';
var sky_100 = exports.sky_100 = '#0c2938';
var slate_10 = exports.slate_10 = '#edf4f7';
var slate_20 = exports.slate_20 = '#cbe3f5';
var slate_30 = exports.slate_30 = '#a7ceeb';
var slate_40 = exports.slate_40 = '#84b7db';
var slate_50 = exports.slate_50 = '#5e95bd';
var slate_60 = exports.slate_60 = '#48799c';
var slate_70 = exports.slate_70 = '#376180';
var slate_80 = exports.slate_80 = '#2d4d63';
var slate_90 = exports.slate_90 = '#243947';
var slate_100 = exports.slate_100 = '#1d2830';
var teal_10 = exports.teal_10 = '#e4f7f6';
var teal_20 = exports.teal_20 = '#a8ede9';
var teal_30 = exports.teal_30 = '#6cd9d2';
var teal_40 = exports.teal_40 = '#45bfb7';
var teal_50 = exports.teal_50 = '#28a199';
var teal_60 = exports.teal_60 = '#17827b';
var teal_70 = exports.teal_70 = '#116963';
var teal_80 = exports.teal_80 = '#0d524d';
var teal_90 = exports.teal_90 = '#0b3d3a';
var teal_100 = exports.teal_100 = '#092b29';
