// mineral-ui-tokens/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _tokens = require('./tokens');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_tokens).default;
  }
});

var _palette = require('./palette');

Object.defineProperty(exports, 'palette', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_palette).default;
  }
});

var _all = require('./all');

Object.keys(_all).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _all[key];
    }
  });
});

var _blue = require('./blue');

Object.defineProperty(exports, 'blue', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_blue).default;
  }
});

var _brand = require('./brand');

Object.defineProperty(exports, 'brand', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_brand).default;
  }
});

var _bronze = require('./bronze');

Object.defineProperty(exports, 'bronze', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_bronze).default;
  }
});

var _dusk = require('./dusk');

Object.defineProperty(exports, 'dusk', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dusk).default;
  }
});

var _gray = require('./gray');

Object.defineProperty(exports, 'gray', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gray).default;
  }
});

var _green = require('./green');

Object.defineProperty(exports, 'green', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_green).default;
  }
});

var _indigo = require('./indigo');

Object.defineProperty(exports, 'indigo', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_indigo).default;
  }
});

var _magenta = require('./magenta');

Object.defineProperty(exports, 'magenta', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_magenta).default;
  }
});

var _purple = require('./purple');

Object.defineProperty(exports, 'purple', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_purple).default;
  }
});

var _red = require('./red');

Object.defineProperty(exports, 'red', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_red).default;
  }
});

var _sky = require('./sky');

Object.defineProperty(exports, 'sky', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sky).default;
  }
});

var _slate = require('./slate');

Object.defineProperty(exports, 'slate', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_slate).default;
  }
});

var _teal = require('./teal');

Object.defineProperty(exports, 'teal', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_teal).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
