// mineral-ui/Flex/FlexItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _memoizeOne = require('memoize-one');

var _memoizeOne2 = _interopRequireDefault(_memoizeOne);

var _styles = require('../styles');

var _Box = require('../Box');

var _Box2 = _interopRequireDefault(_Box);

var _Flex = require('./Flex');

var _Flex2 = _interopRequireDefault(_Flex);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var mapValueToProperty = function mapValueToProperty(property, value) {
  var map = {
    alignSelf: function alignSelf(value) {
      return value === 'start' || value === 'end' ? 'flex-' + value : value;
    },
    flexBasis: function flexBasis(value) {
      return typeof value === 'number' && value < 1 && value !== 0 ? value * 100 + '%' : value;
    },
    flexGrow: function flexGrow(value) {
      return value;
    },
    flexShrink: function flexShrink(value) {
      return value;
    }
  };

  return map[property](value);
};

var styles = function styles(_ref) {
  var alignSelf = _ref.alignSelf,
      breakpoints = _ref.breakpoints,
      grow = _ref.grow,
      shrink = _ref.shrink,
      theme = _ref.theme,
      width = _ref.width;
  return (0, _styles.getResponsiveStyles)({
    breakpoints: breakpoints,
    mapValueToProperty: mapValueToProperty,
    styles: {
      alignSelf: alignSelf,
      flexBasis: width || 'auto',
      flexGrow: grow,
      flexShrink: shrink
    },
    theme: theme
  });
};

var createRootNode = function createRootNode(props) {
  var component = props.flex ? _Flex2.default : _Box2.default;

  return (0, _styles.createStyledComponent)(component, styles, {
    displayName: 'FlexItem',
    filterProps: ['inline', 'width']
  });
};

/**
 * FlexItem is used within [Flex](/components/flex) to lay out other components in
 * your app.
 */

var FlexItem = function (_Component) {
  _inherits(FlexItem, _Component);

  function FlexItem() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, FlexItem);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = FlexItem.__proto__ || Object.getPrototypeOf(FlexItem)).call.apply(_ref2, [this].concat(args))), _this), _this.getRootNode = (0, _memoizeOne2.default)(createRootNode, function (nextProps, prevProps) {
      return nextProps.flex === prevProps.flex;
    }), _temp), _possibleConstructorReturn(_this, _ret);
  }

  // Must be an instance method to avoid affecting other instances memoized keys


  _createClass(FlexItem, [{
    key: 'render',
    value: function render() {
      var Root = this.getRootNode(this.props);

      return _react2.default.createElement(Root, this.props);
    }
  }]);

  return FlexItem;
}(_react.Component);

FlexItem.defaultProps = {
  grow: 0,
  shrink: 1
};
exports.default = FlexItem;
FlexItem.propTypes = process.env.NODE_ENV !== "production" ? {
  /** Align item along the cross axis [[Responsive-capable]](#responsive) */
  alignSelf: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['start']), _propTypes2.default.oneOf(['end']), _propTypes2.default.oneOf(['center']), _propTypes2.default.oneOf(['stretch']), _propTypes2.default.arrayOf(_propTypes2.default.oneOf(['start', 'end', 'center', 'stretch']))]),

  /**
   * Media query (min-width) breakpoints along which to apply props marked
   * "&#xfeff;[[Responsive-capable]](#responsive)&#xfeff;"
   */
  breakpoints: _propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]).isRequired),

  /**
   * When `true`, FlexItem composes [Flex](/components/flex); when `false`, it
   composes [Box](/components/box) */
  flex: _propTypes2.default.bool,

  /**
   * Grow factor along the main axis ([see example](#grow))
   * [[Responsive-capable]](#responsive)
   */
  grow: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number)]),

  /**
   * Shrink factor along the main axis ([see example](#shrink))
   * [[Responsive-capable]](#responsive)
   */
  shrink: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number)])
} : {};
