// mineral-ui/Pagination/PageSizer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Flex = require('../Flex');

var _Form = require('../Form');

var _Select = require('../Select');

var _Select2 = _interopRequireDefault(_Select);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PageSizer = function (_PureComponent) {
  _inherits(PageSizer, _PureComponent);

  function PageSizer() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, PageSizer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = PageSizer.__proto__ || Object.getPrototypeOf(PageSizer)).call.apply(_ref, [this].concat(args))), _this), _this.handleSelect = function (item) {
      var pageSize = parseInt(item.value);
      _this.props.onPageSizeChange(pageSize);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(PageSizer, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          currentPage = _props.currentPage,
          messages = _props.messages,
          pageSize = _props.pageSize,
          pageSizes = _props.pageSizes,
          size = _props.size,
          totalCount = _props.totalCount,
          totalPages = _props.totalPages,
          restProps = _objectWithoutProperties(_props, ['currentPage', 'messages', 'pageSize', 'pageSizes', 'size', 'totalCount', 'totalPages']);

      var rootProps = _extends({}, restProps);

      if (pageSizes.indexOf(pageSize) === -1) {
        throw new Error('pageSizes array does not include pageSize');
      }

      var category = messages.category,
          itemText = messages.itemText,
          status = messages.status;


      var data = pageSizes.map(function (pageSize) {
        return {
          text: itemText(pageSize),
          value: pageSize.toString()
        };
      });

      var first = (currentPage - 1) * pageSize + 1;
      var last = first + pageSize - 1;
      var lastPage = currentPage === totalPages;
      var pageSizerDescription = status(category, first, lastPage ? totalCount : last, totalCount);

      var inputProps = {
        data: data,
        defaultSelectedItem: data.find(function (item) {
          return parseInt(item.value) === pageSize;
        }),
        label: pageSizerDescription,
        hideLabel: true,
        input: _Select2.default,
        caption: pageSizerDescription,
        onChange: this.handleSelect,
        size: size
      };

      return _react2.default.createElement(
        _Flex.FlexItem,
        rootProps,
        _react2.default.createElement(_Form.FormField, inputProps)
      );
    }
  }]);

  return PageSizer;
}(_react.PureComponent);

exports.default = PageSizer;
PageSizer.propTypes = process.env.NODE_ENV !== "production" ? {
  "aria-label": _propTypes2.default.string,
  currentPage: _propTypes2.default.number.isRequired,
  messages: _propTypes2.default.shape({
    category: _propTypes2.default.string.isRequired,
    status: _propTypes2.default.func.isRequired,
    itemText: _propTypes2.default.func.isRequired,
    __exact__: function __exact__(values, prop, displayName) {
      var props = {
        category: true,
        status: true,
        itemText: true
      };
      var extra = [];

      for (var k in values) {
        if (values.hasOwnProperty(k) && !props.hasOwnProperty(k)) {
          extra.push(k);
        }
      }

      if (extra.length > 0) {
        return new Error('Invalid additional prop(s) ' + JSON.stringify(extra));
      }
    }
  }).isRequired,
  onPageSizeChange: _propTypes2.default.func.isRequired,
  pageSize: _propTypes2.default.number.isRequired,
  pageSizes: _propTypes2.default.arrayOf(_propTypes2.default.number.isRequired).isRequired,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large', 'jumbo']),
  totalCount: _propTypes2.default.number.isRequired,
  totalPages: _propTypes2.default.number.isRequired
} : {};
